---
title: "coMethTF: estimating regulatory potential of DNA methylation changes in epigenome-wide association studies"
author:
  - Tiago Chedraoui Silva^[University of Miami Miller School of Medicine, txs902@miami.edu]
  - Lily Wang^[University of Miami Miller School of Medicine, lily.wangg@gmail.com]
output: 
  prettydoc::html_pretty:
    theme: leonids
    toc: true
    df_print: paged
    code_download: false
    toc_depth: 3
bibliography: bibliography.bib    
editor_options:
  chunk_output_type: inline    
vignette: >
    %\VignetteIndexEntry{coMethTF: functional annotation of DMRs and DMS in epigenome-wide association studies}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---
  
<style> body {text-align: justify} </style>
  
```{r settings, include = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
library(dplyr)
library(sesameData)
sesameDataCacheAll()
```

# 1. Introduction

Transcription factors (TFs) are proteins that facilitate the transcription of 
DNA into RNA. A number of recent studies have observed that the binding of TFs 
onto DNA can be affected by DNA methylation, and in turn, DNA methylation can 
also be added or removed by proteins associated with transcription 
factors [@bonder2017disease; @banovich2014methylation; @zhu2016transcription].

To provide functional annotations for differentially methylated regions (DMRs) 
and differentially methylated CpG sites (DMS),`coMethTF` performs integrative 
analyses using matched DNA methylation and gene expression along with 
Transcription Factor Binding Sites (TFBS) data. coMethTF evaluates, prioritizes 
and annotates DNA methylation regions (or sites) with high regulatory potential 
that works synergistically with TFs to regulate target gene expressions, 
without any additional ChIP-seq data. 

The results from `coMethTF` can be used to generate testable hypothesis on the 
synergistic collaboration of DNA methylation changes and TFs in gene regulation.
`coMethTF` can be used either to evaluate regulatory potentials of candidate 
regions or to search for methylation coupled TF regulatory processes in the entire genome.

# 2. Installation 

You can install the development version of the package from GitHub using the following code:

```{r, eval = FALSE}
devtools::install_github("TransBioInfoLab/coMethTF")
```

```{r libs_load, eval = TRUE, include = FALSE}
library(coMethTF)
```

After the package is successfully installed you can load its functions with:

```{r setup, eval = FALSE}
library(coMethTF)
```

# 3. Dataset

The input of `coMethTF` are DNA methylation and gene expression datasets measured 
on the same set of samples. Using additional information from the 
HOCOMOCO database [@kulakovskiy2016hocomoco], we will then create a data frame 
with information on the triplet: DMR (or CpG), TF and target gene. 

For illustration, we will use chromosome 21 data from 38 TCGA-COAD samples.

## 3.1 DNA methylation dataset

The DNA methylation dataset is a matrix with methylation beta or M-values. 
If there are potential confounding factors (e.g. batch effect, age, sex) in the dataset, 
this matrix can also contain residuals from fitting linear regression 
instead (see **Section 5** below). 

The samples are in the columns and regions/probes are in the rows.   

### 3.1.1 Analysis for individual CpGs data

The individual CpGs data can be, for example, differentially methylated sites 
(DMS) obtained in a DNA methylation study.

```{R warning=FALSE}
dna.met.chr21 <- get(data("dna.met.chr21"))
```

```{R}
dna.met.chr21[1:5,1:5]
```

```{R}
dna.met.chr21 <- map_probes_to_regions(dna.met.chr21)
```


```{R}
dna.met.chr21[1:5,1:5]
```

### 3.1.2 DMRs data

Differentially Methylated Regions (DMRs) associated with phenotypes such as tumor stage 
can be obtained from R packages such as \code{coMethDMR}, \code{comb-p}, \code{DMRcate} 
and many others. The methylation levels in multiple CpGs within the DMRs need to be 
summarized (e.g. using medians), then the analysis for DMR will proceed in the same way 
as those for CpGs. See **Section 6** "Identifying co-methylated clusters and summarizing 
regions using medians for DMRs" below for more details.   


## 3.2 Gene expression dataset

The gene expression dataset is a matrix with log2 transformed and normalized gene expression values. 
If there are potential confounding factors (e.g. batch effect, age, sex) in the dataset, 
this matrix can also contain residuals from linear regression instead (see **Section 5** below).
The samples are in the columns and the genes are in the rows. 

```{R}
data("gene.exp.chr21", package = "coMethTF")
gene.exp.chr21[1:5,1:5]
```

## 3.3 Creating triplet dataset
### Mapping DMR region (or CpG) to the target gene

The function `get_region_target_gene` provides two methods to link a DMR to 
a target gene, by (1) mapping the region to the closest gene (`method = "closest.gene"`) or 
(2) mapping the region to all the genes within a window 
(default size = 500Kbp around the region, i.e. +- 250K bp from start or end of the region) (`method = "window"`).

For promoter analysis, we recommend setting `method = "closest.gene"`.
For analysis of distal regions, 
we recommend setting  `method = "window"`. 
Note that the distal analysis will be more time and resource consuming. 
 
```{R}
# Get CpG information
probes.gr <- make_granges_from_names(rownames(dna.met.chr21))
class(probes.gr)
```

```{R}
# Map probes to nearest gene
region.target.closest <- get_region_target_gene(
    regions.gr = probes.gr, 
    genome = "hg38", 
    method = "closest.gene"
)
dim(region.target.closest)
region.target.closest
```

```{R}
# Map probes to genes within 500kb window
region.target.window <- get_region_target_gene(
    regions.gr = probes.gr, 
    genome = "hg38", 
    method = "window"
)

dim(region.target.window)
region.target.window
```

Some of the columns of the data frame output from `get_region_target_gene` 
when the method is set to `"window"` are described below:

* regionID: chromossome, start, end
* regionID.extended: same as regionID, but start and end values of the regions will be 
regionID start - window.width/2 and  regionID end + window.width/2 
* window.extended.width:  width of window around region used to find target genes
* Distance.region.gene: distance from region to target gene TSS


### Mapping region to TF

JASPAR 2020 is an open-access database of curated, non-redundant transcription 
factor (TF)-binding profiles [@JASPAR2020, @fornes2020jaspar], which contains 
more the 500 human TF motifs.

The function `get_tf_in_region` uses `motifmatchr` [@motifmatchr] to scan them for occurrences of motifs.
The argument `window.size` will be use to extend the region to scan
for the motifs, for example, a `window.size` of `50` will add `25` bp 
upstream and `25` bp downstream of the original region.

```{R, results = "hide"}
region.tf <- get_tf_in_region(
    region = probes.gr, 
    genome = "hg38", 
    window.size = 50,
    p.cutoff = 5e-5
)
```

```{R}
head(region.tf)
dim(region.tf)
```

### Triplet data frame â€“ matching DMR (or CpGs) with TFs and target genes

The triplet is a data frame with the following columns:

* `target`: gene identifier (obtained from row names of the gene expression matrix),
* `regionID`: region/cpg identifier (obtained from row names of the DNA methylation matrix)
* `TF`: gene identifier (obtained from the row names of the gene expression matrix)

```{R, message = FALSE, warning = FALSE}
# Merging the results from the previous sections (region and target, and region and TF)
triplet <- dplyr::inner_join(region.target.closest, region.tf)
```

```{R}
head(triplet)
```

There may also be multiple rows for a region, when multiple target gene and/or TFs are found close to it. 

# 4. Evaluating the regulatory potential of DMRs (or CpGs)

Because TF binding to DNA can be influenced by (or influence) DNA methylation levels nearby [@yin2017impact], 
target gene expression can often be influenced by the synergistic effects of both TF and DNA methylation. 
In other words, TF activities in gene regulation is affected by DNA methylation or vice versa. 

Our goal then is to highlight DNA methylation regions (or CpGs) where these synergistic DNAm and TF collaborations occur. 
We will perform analyses using the 3 datasets described above in Section 3:

* DNA methylation matrix
* Gene expression matrix
* The triplet data frame with DNA methylation region, TF, and target gene

## 4.1 Analysis using model with methylation by TF interaction

The function `interaction_model` assess the regulatory impact of DNA methylation on TF regulation of target genes via two approaches: 

**Model 1** (considering DNAm values as a continuous variable) - we fit a model with TF by DNA methylation interaction using all samples.

$$log_2(RNA target)  \sim log_2(TF) + DNAm + log_2(TF) * DNAm$$

**Model 2** (considering DNAm values as a binary variable) - we define a binary variable 
`DNAm Group` for DNA methylation values (high = 1, low = 0). That is, samples with the 
highest DNAm levels (top 25 percent) has high = 1, samples with lowest DNAm levels (bottom 25 pecent) has high = 0.  

Note that in this implementation, only samples wih DNAm values in the first and last quartiles are considered.

$$log_2(RNA target) \sim log_2(TF) + \text{DNAm Group} + log_2(TF) * \text{DNAm Group}$$
 
```{R interaction_model, message = FALSE, results = "hide"}
results <- interaction_model(
    triplet = triplet, 
    dnam = dna.met.chr21,
    exp = gene.exp.chr21
)
```

The output of `interaction_model` function will be a data frame with the following variables:

* `pval_<variable>`: p-value for a tested variable (methylation or TF), given the other variables included in the model.
* `estimate_<variable>`: estimated effect for a variable. If estimate > 0, increasing values 
of the variable corresponds to increased outcome values (target gene expression). 
If estimate < 0, increasing values of the variable correspond to decreased target gene expression levels.


The following columns are provided for the results of fitting **Model 1** to triplet data:

* `pval_met`: p-value for DNA methylation variable
* `pval_rna.tf` : p-value for TF expression
* `pval_met.rna.tf`: : p-value for DNA methylation by  TF interaction
* `estimate_met`: estimate for DNA methylation variable
* `estimate_rna.tf`: estimate  for TF expression
* `estimate_met.rna.tf`: estimate for DNA methylation by  TF interaction

```{R interaction_model_results}
dim(results)
colnames(results)

# Results for full model (with all samples)
results %>% dplyr::select(
    c(1,4,5,grep("quant",colnames(results),invert = TRUE))
    ) %>% head
```


The following columns are provided for the results of fitting **Model 2** (quartile model) to triplet data:

* `quant_pval_metGrp`: p-value for DNA methylation variable
* `quant_pval_rna.tf` : p-value for TF expression
* `quant_pval_metGrp:rna.tf`: : p-value for DNA methylation by  TF interaction
* `quant_estimates_metGrp`: estimate for DNA methylation variable
* `quant_estimates_rna.tf`: estimate  for TF expression
* `quant_estimates_metGrp:rna.tf`: estimate for DNA methylation by  TF interaction


```{R}
# Results for quartile model
results %>% dplyr::select(
  c(1,4,5,grep("quant",colnames(results)))
  ) %>% head
```

To select significant DNAm by TF intereaction from both models, you can subset the results
using the p-value from each linear model.

```{r, include = FALSE, eval = TRUE}
results.sig <- results %>% dplyr::filter(`pval_met:rna.tf` < 0.05 & `quant_pval_metGrp:rna.tf` < 0.05)
```

```{R}
head(results.sig)
```

## 4.2 Stratified analysis by high and low DNA methylation levels

For significant triplets that are identified above, we can further assess how 
gene regulation by TF changes when DNAm is high or low. To this end, the function 
`stratified_model` fits two separate models (see below) to only the 
samples in highest DNAm levels (top 25 percent), and then to only the samples in 
lowest DNAm levels (bottom 25 percent), separately.

$$\text{Stratified Model: } log_2(RNA target) \sim log_2(TF)$$

```{R stratified_model, message = FALSE, results = "hide"}
results.stratified.model <- stratified_model(
    triplet = triplet, 
    dnam = dna.met.chr21,
    exp = gene.exp.chr21
)
```

```{R}
results.stratified.model %>% head
```

## 4.3 Visualization of data  

The function `plot_interaction_model` will create figures to visualize the data, 
in a way that corresponds to the linear model we considered above. 
It requires the output from the function `interaction_model` (a dataframe), 
the DNA methylation matrix and the gene expression matrix as input. 


```{R, plot_stratified_model, message = FALSE, results = "hide", eval = TRUE}
plots <- plot_stratified_model(
    triplet.results = results.stratified.model[1,], 
    dnam = dna.met.chr21, 
    exp = gene.exp.chr21
)
```

```{R, fig.height = 14, fig.width = 12, eval = TRUE}
plots[[1]]
```

The first row of the figures shows pairwise associations between DNA methylation, 
TF and target gene expression levels. 

The second row of the figures show how much TF activity on target 
gene expression levels vary by DNA methylation levels. When TF by methylation 
interaction is significant (Section 4.1), we expect the association between TF 
and target gene expression vary depending on whether DNA methylation is low or high. 

In this example, when DNA methylation is low, more abundant TF corresponds to 
decreased gene expression (a repressor TF). On the other hand, when DNA methylation 
level is high, target gene expression is relatively independent of the amount of TF available. 
One possibility is that DNA methylation might affect TF binding in this case. 

If you want to highlight different group of samples, you can also use the metadata argument
with a dataframe with rownames as samples (same order as the columns of the 
DNAm and gene expression), and a column highlighting the groups.

```{R, message = TRUE, results = "hide"}
data("clinical")
metadata <- clinical[,c("sample_type"),drop = FALSE]
```

```{R, eval = TRUE, results = "hide", message = FALSE}
plots <- plot_stratified_model(
        triplet.results = results.stratified.model[1,,drop = FALSE], 
        dnam = dna.met.chr21, 
        exp = gene.exp.chr21,
        metadata = metadata
)
```


```{R, fig.height = 14, fig.width = 12, eval = FALSE}
plots[[1]]
```

# 5. Results interpretation

Shown below are some expected results from fitting Models 1 & 2 described in 
**Section 4.1** above, depending on TF binding preferences. Please note that 
there can be more possible scenarios than those listed here, therefore, 
careful evaluation of the statistical models and visualizations of data as 
described in **Section 4** is needed to gain a good understanding of the 
multi-omics data. 

![Illustration - Possible scenarios](scenarios.png)

# 6. Identifying co-methylated clusters and summarizing regions using medians for DMRs

Given regions for DMRs, the R package [coMethDMR](https://github.com/TransBioInfoLab/coMethDMR) 
can be used to identify co-methylated regions [@gomez2019comethdmr]. 
To install it, please use the following command:

```{R, eval = FALSE}
devtools::install_github("TransBioInfoLab/coMethDMR")
```

We can then summarize methylation levels from multiple CpGs within each region using median. 

```{r, eval = TRUE, include = TRUE, message = FALSE, results = "hide"}
library(coMethDMR)
library(dplyr)
```

For illustration, we will use probe clusters in predefined genomic regions 
from the `coMethDMR` R package here. 

```{R cluster_probes, eval = TRUE}
probes.cluster <- coMethDMR::getPredefinedCluster(
  arrayType = "450k",
  clusterType = "regions"
)
```

```{R}
message("Number of clusters: ",length(probes.cluster))
probes.cluster %>% head
```

For real data analyis, please replace `probes.cluster` with probes in DMRs 
identified in epigenomewide association study. The  `coMethDMR::GetCpGsInAllRegion` 
function can be used to extract probes within the DMRs.  


To find co-methylated regions within the DMRs, first we need to extract probes within the DMRs.  
```{r, eval = TRUE}
probes.cluster <- coMethDMR::GetCpGsInAllRegion(
    regionName_char = c("chr21:37432493-37432782",
                        "chr21:18884682-18885210", 
                        "chr21:27107799-27107982"),
      arrayType = "450k"
)
probes.cluster
```


Next we find co-methylated regions within the DMRs.
```{R cometh, eval = TRUE}
data("dna.met.chr21")
# just keep the used probes
dna.met.chr21 <- dna.met.chr21[rownames(dna.met.chr21) %in% unique(unlist(probes.cluster)),]
```

```{R,eval = TRUE}
coMeth_ls <- coMethDMR::CoMethAllRegions(
    dnam = dna.met.chr21,
    betaToM = TRUE,
    method = "spearman",
    minPairwiseCorr = 0.2,
    CpGs_ls = probes.cluster,
    arrayType = "450k",
    returnAllCpGs = FALSE,
    output = "CpGs",
    nCores_int = 1
)
```


```{R}
message("Number of co-methylated cluster: ",length(coMeth_ls))
coMeth_ls %>% head
```

Finally, we compute median methylation of co-methylated regions. 

```{R, results = "hide"}
median.df.tumor <- plyr::ldply(
    coMeth_ls[!duplicated(names(coMeth_ls))],
    function(probes){
        matrixStats::colMedians(dna.met.chr21[probes,])
        }, .progress = "time",.parallel = FALSE
)
```

```{R}
rownames(median.df.tumor) <- median.df.tumor$.id
median.df.tumor$.id <- NULL
colnames(median.df.tumor) <- colnames(dna.met.chr21)
dim(median.df.tumor)
head(median.df.tumor)
```

The analysis described in **Section 3.3** and **Section 4** can then be performed 
for DMRs using `median.df.tumor`, in the same way as those for dataset `dna.met.chr21`. 

# 7. Controlling effects from confounding variables

Both gene expressions and DNA methylation levels can be affected by age, sex, 
shifting in cell types, batch effects and other confounding variables.  
In this section, we illustrate analysis workflow that reduces confounding effects, 
by first extracting the residual data with the function 
`get_residuals`, before fitting the models discussed above in Section 4. 

The `get_residuals` function will use gene expression (or DNA methylation data) 
and phenotype data as input. 
To remove confounding effects in gene expression data, we use the `get_residuals` 
function which extract residuals after fitting the following model: 
$$features \sim covariate_{1} + covariate_{2} ... + covariate_{N}$$


```{R residuals, results = "hide", eval = TRUE}
data("gene.exp.chr21")
data("clinical")
metadata <- clinical[,c("sample_type","gender")]

gene.exp.chr21.residuals <- get_residuals(gene.exp.chr21[1:250,], metadata) %>% as.matrix()
```

```{R, eval = TRUE}
gene.exp.chr21.residuals[1:5,1:5]
```

```{R, results = "hide", eval = TRUE}
data("dna.met.chr21")
dna.met.chr21 <- map_probes_to_regions(dna.met.chr21)
dna.met.chr21.residuals <- get_residuals(dna.met.chr21, metadata) %>% as.matrix()
```

```{R, eval = TRUE}
dna.met.chr21.residuals[1:5,1:5]
```

The `interaction_model` function can then be applied to these residuals data, 
as the following:  `rna.target ~ met + rna.tf + rna.tf * met`.


```{R, message = FALSE, results = "hide", eval = TRUE}
results <- interaction_model(
    triplet = triplet, 
    dnam = dna.met.chr21.residuals, 
    exp = gene.exp.chr21.residuals
)
```

# 8. Session information
```{R,size = 'tiny'}
sessionInfo()
```

# 9. Bibliography
