---
title: "coMethTF: functional annotation of DMRs identified in epigenome-wide association studies"
output:
  rmarkdown::html_document:
    theme: lumen
    toc: true
    number_sections: true
    df_print: paged
    code_download: false
    toc_float:
      collapsed: yes
    toc_depth: 3
bibliography: bibliography.bib    
editor_options:
  chunk_output_type: inline    
vignette: >
  %\VignetteIndexEntry{How to use MethTF}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<style> body {text-align: justify} </style>

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction

## Description

Transcription factors (TFs) are proteins that facilitate the transcription of DNA into RNA. A number of recent studies have observed that the binding of transcription factor onto DNA can be affected by DNA methylation, and in turn, DNA methylation can also be added or removed by proteins associated with transcription factors [@bonder2017disease; @banovich2014methylation; @zhu2016transcription]. Using methylation-sensitive SELEX (systematic evolution of ligands by exponential enrichment), @yin2017impact classified 519 TFs into several categories â€“ TFs whose binding strength increased, decreased, or not affected by DNA methylation, as well as those not containing CpGs in their binding motifs. 

To determine whether TF regulatory activity is enhanced or reduced by differentially methylated regions (DMRs), the gold standard would be to perform ChIP-Seq experiment for all candidate transcription factors with binding sites close to the DMRs. However, information on ChIP-Seq experiment for disease relevant tissues are often not available, and because experimentally determining DNA methylation at TFBS can be time and resource consuming, computationally approaches are often used to prioritize active TFs in DNA methylation studies. 

Here we present `coMethTF`, which performs integrative analyses on all three key components in gene regulation, that is,  DNA methylation, gene expression, and TFBS data, to provide functional annotations for DMRs. The results from coMethTF can be used to generate testable hypothesis on the synergistic interaction of DMRs and TF in gene regulation.  `coMethTF` is a computational tool that evaluates, prioritizes and annotates DNA methylation regions with high regulatory potential using matched DNA methylation and gene expression data, without any additional ChIP-seq data. `coMethTF` can be used either to evaluate regulatory potentials of candidate regions or to search for methylation coupled TF regulatory processes in the entire genome. 

## Install 

You can install the development version of the package from GitHub using the following code:

```{r, eval = FALSE}
devtools::install_github("TransBioInfoLab/coMethTF")
```

## Load

After the package is successfully installed you can load its functions with `library` command.

```{r setup}
library(coMethTF)
```

# Interaction model

## Data

```{R, message = FALSE}
library(dplyr)
```

```{R}
triplet <- data.frame("regionID" = paste0("region_",1:9),
                       "TF" = paste0("ESNG",10),
                       "target" = paste0("ESNG",1:9))
dnam <- runif(200) %>% matrix(10)
colnames(dnam) <- paste0("Sample_",1:20)
rownames(dnam) <- paste0("region_",1:10)
exp <- rexp(200) %>% matrix(10)
colnames(exp) <- paste0("Sample_",1:20)
rownames(exp) <- paste0("ESNG",1:10)

head(triplet)
dnam[1:5,1:5]
exp[1:5,1:5]
```


## Perform interaction model

The function `interaction_model` will evaluate the response of the target gene expression
using as predictor the DNA methylation level and the TF expression: `log2(rna.target + 1) ~ met + log2(rna.tf + 1) + log2(rna.tf + 1) * met`

```{R, message = FALSE, results = "hide"}
results <- interaction_model(triplet, dnam, exp)
```

The output of `interaction_model` function will be a data frame with the 
following informations:

* variable `p-values`: p-value for the t-statistic of the hypothesis test that the corresponding coefficient is equal to zero or not. If the p-value > 0.05 the term is not significant given the other terms in the model.
* variable `estimates`: coefficient estimate. If estimate > 0, the increase of the variable 
will increase the output value (target gene expression). If estimate < 0, the increase of the variable 
will decrease the target gene expression.

The following columns are provided:

* `pval_met`: p-value for DNA methylation variable
* `pval_rna.tf` : p-value for TF expression
* `pval_met.rna.tf`: : p-value for DNA methylation and TF interaction
* `estimate_met`: estimate for DNA methylation variable
* `estimate_rna.tf`: estimate  for TF expression
* `estimate_met.rna.tf`: estimate for DNA methylation and TF interaction

```{R}
results
```

## Visualization of interaction

The function `plot_interaction_model` will create a plot to visualize the data and results.
It requires as input the data frame output from the function `interaction_model`, 
the DNA methylation matrix and the gene expression matrix.


```{R, message = FALSE, results = "hide"}
plots <- plot_interaction_model(results, dnam, exp)
```

```{R, fig.height = 9, fig.width = 9}
plots[[1]]
```

# Use case

# Session information
```{R}
devtools::session_info()
```

# Bibliography
