---
title: "coMethTF: functional annotation of DMRs identified in epigenome-wide association studies"
author:
  - Tiago Chedraoui Silva^[Miami University, txs902@miami.edu]
  - Lily wang^[Miami University]
output:
  rmarkdown::html_document:
    theme: lumen
    toc: true
    number_sections: true
    df_print: paged
    code_download: false
    toc_float:
      collapsed: yes
    toc_depth: 3
bibliography: bibliography.bib    
editor_options:
  chunk_output_type: inline    
vignette: >
    %\VignetteIndexEntry{How to use MethTF}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---
  
<style> body {text-align: justify} </style>
  
```{r, include = FALSE}
knitr::opts_chunk$set(
collapse = TRUE,
comment = "#>"
)
library(dplyr)
```

# Introduction

## Description

Transcription factors (TFs) are proteins that facilitate the transcription of DNA into RNA. A number of recent studies have observed that the binding of transcription factor onto DNA can be affected by DNA methylation, and in turn, DNA methylation can also be added or removed by proteins associated with transcription factors [@bonder2017disease; @banovich2014methylation; @zhu2016transcription]. Using methylation-sensitive SELEX (systematic evolution of ligands by exponential enrichment), @yin2017impact classified 519 TFs into several categories â€“ TFs whose binding strength increased, decreased, or not affected by DNA methylation, as well as those not containing CpGs in their binding motifs. 

To determine whether TF regulatory activity is enhanced or reduced by differentially methylated regions (DMRs), the gold standard would be to perform ChIP-Seq experiment for all candidate transcription factors with binding sites close to the DMRs. However, information on ChIP-Seq experiment for disease relevant tissues are often not available, and because experimentally determining DNA methylation at TFBS can be time and resource consuming, computationally approaches are often used to prioritize active TFs in DNA methylation studies. 

Here we present `coMethTF`, which performs integrative analyses on all three key components in gene regulation, that is,  DNA methylation, gene expression, and TFBS data, to provide functional annotations for DMRs. The results from coMethTF can be used to generate testable hypothesis on the synergistic interaction of DMRs and TF in gene regulation. `coMethTF` is a computational tool that evaluates, prioritizes and annotates DNA methylation regions with high regulatory potential using matched DNA methylation and gene expression data, without any additional ChIP-seq data. `coMethTF` can be used either to evaluate regulatory potentials of candidate regions or to search for methylation coupled TF regulatory processes in the entire genome. 

## Install 

You can install the development version of the package from GitHub using the following code:

```{r, eval = FALSE}
devtools::install_github("TransBioInfoLab/coMethTF")
```

## Load

```{r, eval = TRUE, include = FALSE}
devtools::load_all(".")
```

After the package is successfully installed you can load its functions with `library` command.

```{r setup, eval = FALSE}
library(coMethTF)
```

# Interaction analysis

## Data

To perform the interaction analysis there are 3 main required data:

* DNA methylation matrix
* Gene expression matrix
* A data frame with the Triplet region, TF, and target gene

The next sections will provide an example of DNA methylation and gene expression matrix 
and how you can get the information required to create the triplet data frame.

### DNA methylation

The DNA methylation data should be a matrix with samples as columns and regions/probes 
as rows. 

```{R}
data("dna.met.chr21")
dna.met.chr21[1:5,1:5]
dna.met.chr21 <- map_probes_to_regions(dna.met.chr21)
dna.met.chr21[1:5,1:5]
```

### Gene expression

The expected gene expression data is a matrix with samples in columns and 
genes in the rows.

```{R}
data("gene.exp.chr21")
gene.exp.chr21[1:5,1:5]
```

## Mapping region to target gene

The function `get_region_target_gene` provides two methods to map a region to 
a target gene. One that maps the region to the closest gene (`method = "closest.gene"`) and one that maps 
the region to all the genes within a windown (default size = 500Kbp around the region) 
 (`method = "window"`) .
 
```{R}
# Get CpG information
library(sesame)
HM450.hg38.manifest <- sesameDataGet("HM450.hg38.manifest")
probes.gr <- make_granges_from_names(rownames(dna.met.chr21))
class(probes.gr)

# Map probes to nearest gene
mapping.closest <- get_region_target_gene(probes.gr, genome = "hg38", method = "closest.gene")
dim(mapping.closest)
mapping.closest

# Map probes to genes within 500kb window
mapping.window <- get_region_target_gene(probes.gr, genome = "hg38", method = "window")
dim(mapping.window)
mapping.window
```

## Mapping region to TF


The function `get_tf_in_region` maps the regions to TF using a precomputed dataset using
HOCOMOCO motifs [@kulakovskiy2016hocomoco]. For each motif we scanned 500 bp around each EPIC and 
HM450 arrays probes. Then for each motif found in a region we extract TF within the same Tf family or subfamily,
using TFClass [@wingender2013tfclass], since they have similar binding motifs.

```{R, results = "hide"}
# we will only map the first 10 regions, since this function may take a while to compute
region.tf <- get_tf_in_region(probes.gr[1:20], genome = "hg38", arrayType = "450k", classification = "subfamily")
```

```{R}
head(region.tf)
```


## Triplet: TF, target gene, region

The triple is a data frame with the following columns:

* `target`: gene identifier within the row names of the gene expression matrix, 
* `regionID`: region/cpg identifier row names of the DNA methylation matrix  
* `TF`: gene identifier within the row names of the gene expression matrix

```{R}
# Merging the results from the previous sections (region and target, and region and TF)
triplet <- dplyr::inner_join(
    mapping.closest[,c("regionID","ensembl_gene_id")],
    region.tf[,c("regionID","TF_ensembl_gene_id")]
    )
colnames(triplet) <- c("regionID","target","TF")
head(triplet)
```


## Evaluating DNA methylation and TF interaction

The function `interaction_model` will evaluate the response of the target gene expression
using as predictor the DNA methylation level and the TF expression: 
`log2(rna.target + 1) ~ met + log2(rna.tf + 1) + log2(rna.tf + 1) * met`

Also, it will fit a separate model to samples with  **DNA methylation in low quartile** 
and another model to samples with  **DNA methylation in high quartile**.


```{R, message = FALSE, results = "hide"}
results <- interaction_model(triplet, dna.met.chr21, gene.exp.chr21)
```

The output of `interaction_model` function will be a data frame with the 
following informations:

* variable `p-values`: p-value for the t-statistic of the hypothesis test that the corresponding coefficient is equal to zero or not. If the p-value > 0.05 the term is not significant given the other terms in the model.
* variable `estimates`: coefficient estimate. If estimate > 0, the increase of the variable 
will increase the output value (target gene expression). If estimate < 0, the increase of the variable 
will decrease the target gene expression.

The following columns are provided:

* `pval_met`: p-value for DNA methylation variable
* `pval_rna.tf` : p-value for TF expression
* `pval_met.rna.tf`: : p-value for DNA methylation and TF interaction
* `estimate_met`: estimate for DNA methylation variable
* `estimate_rna.tf`: estimate  for TF expression
* `estimate_met.rna.tf`: estimate for DNA methylation and TF interaction

The results for the separated model fit to samples with **DNA methylation in low quartile** are provided within following columns:

* `DNAmlow_pval_met`     
* `DNAmlow_pval_rna.tf`
* `DNAmlow_pval_met.rna.tf`      
* `DNAmlow_estimate_met`         
* `DNAmlow_estimate_rna.tf` 
* `DNAmlow_estimate_met.rna.tf`  

The results for the separated model fit to samples with  **DNA methylation in low quartile** are provided within following columns:

* `DNAmhigh_pval_met`            
* `DNAmhigh_pval_rna.tf`         
* `DNAmhigh_pval_met.rna.tf`    
* `DNAmhigh_estimate_met`        
* `DNAmhigh_estimate_rna.tf`     
* `DNAmhigh_estimate_met.rna.tf`

```{R}
dim(results)
colnames(results)

# Results for model with all samples
results %>% dplyr::select(c(1,4,5,grep("DNAmlow|DNAmhigh",colnames(results),invert = TRUE)))

# Results for DNA methylation in low quartile model
results %>% dplyr::select(c(1,4,5,grep("DNAmlow",colnames(results))))

# Results for DNA methylation in high quartile model
results %>% dplyr::select(c(1,4,5,grep("DNAmhigh",colnames(results))))
```

## Visualization of interaction

The function `plot_interaction_model` will create a plot to visualize the data and results.
It requires as input the data frame output from the function `interaction_model`, 
the DNA methylation matrix and the gene expression matrix.


```{R, message = FALSE, results = "hide"}
plots <- plot_interaction_model(results, dna.met.chr21, gene.exp.chr21)
```

```{R, fig.height = 14, fig.width = 12}
plots[[1]]
```


If you want to highlight different group of samples, you can also use the metadata argument
with a dataframe with rownames as samples (same order as the columns of the DNAm and gene expression),
and a column highlighting the groups.


```{R, message = FALSE, results = "hide"}
metadata <- data.frame(row.names = colnames(dna.met.chr21),group = c(rep("Tumor",25),rep("Normal",25)))
plots <- plot_interaction_model(results, dna.met.chr21, gene.exp.chr21, metadata)
```

```{R, fig.height = 14, fig.width = 12}
plots[[1]]
```

## Controlling confounding effects

You can control for confounding effects using the residual data with the function
`get_residuals`. This function will use metadata to adjust the data for 
confounding effects as follows: features ~ covariate1 + covariate2 ... + covariateN.

```{R, results = "hide"}
data("gene.exp.chr21")
metadata <- data.frame(
     row.names = colnames(gene.exp.chr21),
     sex = c(rep("Male",25),rep("Female",25)),
     age = sample(50:90, 50, replace = TRUE)
)
gene.exp.chr21.residuals <- get_residuals(gene.exp.chr21, metadata)
```
```{R}
gene.exp.chr21.residuals[1:5,1:5]
```

```{R, results = "hide"}
data("dna.met.chr21")
dna.met.chr21 <- map_probes_to_regions(dna.met.chr21)
dna.met.chr21.residuals <- get_residuals(dna.met.chr21,metadata)
```

```{R}
dna.met.chr21.residuals[1:5,1:5]
```

These residuals data can then be used with the `interaction_model` function.
which will evaluate the response of the target gene expression
using as predictor the DNA methylation level and the TF expression as follows: 
`rna.target ~ met + rna.tf + rna.tf * met`.


```{R, message = FALSE, results = "hide"}
results <- interaction_model(triplet, dna.met.chr21.residuals, gene.exp.chr21.residuals)
```

# Session information
```{R}
devtools::session_info()
```

# Bibliography
