<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Assessing the regulatory potential of DNA methylation regions or sites on gene transcription â€¢ MethReg</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Assessing the regulatory potential of DNA methylation regions or sites on gene transcription">
<meta property="og:description" content="MethReg evaluates, annotates and prioritizes DNA methylation regions 
  (or sites) with high regulatory potential, which work synergistically with transcription
  factors to regulate target gene expressions, without any additional ChIP-seq data.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">MethReg</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/MethReg.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="methreg" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#methreg" class="anchor"></a>MethReg</h1></div>
<!-- badges: start -->

<p><code>MethReg</code> can be used to generate testable hypothesis on the synergistic interaction of DMRs and TFs in gene regulation. <code>MethReg</code> can be used either to evaluate regulatory potentials of candidate regions or to search for methylation coupled TF regulatory processes in the entire genome.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install the released version of MethReg from GitHub with:</p>
<div class="sourceCode" id="cb1"><pre class="r"><span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"TransBioInfoLab/MethReg"</span>)</pre></div>
</div>
<div id="example" class="section level2">
<h2 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h2>
<p>This is a basic example which shows you how to use the package:</p>
<div class="sourceCode" id="cb2"><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">MethReg</span>)
<span class="co">#---------------------------------------</span>
<span class="co"># Data input</span>
<span class="co">#---------------------------------------</span>
<span class="co"># 1) Gene expression matrix</span>
<span class="co"># 2) DNA methylation</span>
<span class="co"># With same column names</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"dna.met.chr21"</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"gene.exp.chr21.log2"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">dna.met.chr21</span>) <span class="kw">==</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">gene.exp.chr21.log2</span>))
<span class="co">#&gt; [1] TRUE</span>

<span class="co"># Since we are working with regions we need to map our 450k array to regions</span>
<span class="no">dna.met.chr21</span> <span class="kw">&lt;-</span> <span class="fu"><a href="reference/make_se_from_dnam_probes.html">make_se_from_dnam_probes</a></span>(<span class="no">dna.met.chr21</span>)
<span class="co">#&gt; o Creating a SummarizedExperiment from DNA methylation input</span>
<span class="co">#&gt; oo Fetching probes metadata</span>
<span class="co">#&gt; oo Removing masked probes</span>
<span class="co">#&gt; oo Preparing SummarizedExperiment object</span></pre></div>
<div class="sourceCode" id="cb3"><pre class="r"><span class="co">#---------------------------------------</span>
<span class="co"># Mapping regions</span>
<span class="co">#---------------------------------------</span>
<span class="co"># For each region get target gene and predicted TF biding to the regions</span>
<span class="co"># get_triplet incorporates two other functions:</span>
<span class="co"># 1) get_region_target_gene</span>
<span class="co"># 2) get_tf_in_region</span>
<span class="no">triplet</span> <span class="kw">&lt;-</span> <span class="fu"><a href="reference/create_triplet_distance_based.html">create_triplet_distance_based</a></span>(
    <span class="kw">region</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">dna.met.chr21</span>),
    <span class="kw">motif.search.window.size</span> <span class="kw">=</span> <span class="fl">50</span>,
    <span class="kw">motif.search.p.cutoff</span> <span class="kw">=</span> <span class="fl">10</span>^-<span class="fl">3</span>,
    <span class="kw">target.method</span> <span class="kw">=</span> <span class="st">"genes.promoter.overlap"</span>,
    <span class="kw">genome</span> <span class="kw">=</span> <span class="st">"hg19"</span>,
    <span class="kw">cores</span> <span class="kw">=</span> <span class="fl">1</span>
)
<span class="co">#&gt; Finding target genes</span>
<span class="co">#&gt; Mapping regions to the closest gene</span>
<span class="co">#&gt; Looking for TFBS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Evaluating 554 JASPAR Human TF motifs</span>
<span class="co">#&gt; This may take a while...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'S4Vectors'</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     expand.grid</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'Biostrings'</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     strsplit</span>
<span class="co">#&gt; Preparing output</span>
<span class="co">#&gt; Joining, by = "regionID"</span></pre></div>
<div class="sourceCode" id="cb4"><pre class="r"><span class="co">#---------------------------------------</span>
<span class="co"># Evaluate two models: </span>
<span class="co">#---------------------------------------</span>
<span class="co"># 1) target gene ~ TF + DNAm + TF * DNAm</span>
<span class="co"># 2) target gene ~ TF + DNAm_group + TF * DNAm_group </span>
<span class="co"># where DNAm_group is a binary indicator if the sample belongs to: Q4 or Q1</span>
<span class="no">results</span> <span class="kw">&lt;-</span> <span class="fu"><a href="reference/interaction_model.html">interaction_model</a></span>(
    <span class="kw">triplet</span> <span class="kw">=</span> <span class="no">triplet</span>,
    <span class="kw">dnam</span> <span class="kw">=</span> <span class="no">dna.met.chr21</span>,
    <span class="kw">exp</span> <span class="kw">=</span> <span class="no">gene.exp.chr21.log2</span>
)
<span class="co">#&gt; Removing genes with RNA expression equal to 0 for all samples from triplets</span>
<span class="co">#&gt; Removing triplet with no DNA methylation information for more than 25% of the samples</span>
<span class="co">#&gt; Filtering results to have interaction, TF or DNAm significant</span>
<span class="co">#&gt; Filtering results to wilcoxon test TF Q1 vs Q4 not significant</span></pre></div>
<div class="sourceCode" id="cb5"><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">results</span>)
<span class="co">#&gt;                  regionID target_gene_name          target</span>
<span class="co">#&gt; 1 chr21:30372219-30372220          RPL23P2 ENSG00000176054</span>
<span class="co">#&gt; 2 chr21:30430511-30430512       AF129075.5 ENSG00000231125</span>
<span class="co">#&gt; 3 chr21:33109780-33109781       AP000255.6 ENSG00000273091</span>
<span class="co">#&gt; 4 chr21:40692859-40692860            BRWD1 ENSG00000185658</span>
<span class="co">#&gt; 5 chr21:43983587-43983588       AP001625.6 ENSG00000235772</span>
<span class="co">#&gt; 6 chr21:45468403-45468404          H2AFZP1 ENSG00000213440</span>
<span class="co">#&gt;   TF_external_gene_name              TF TF_symbol target_symbol   pval_met</span>
<span class="co">#&gt; 1                  ETS2 ENSG00000157557      ETS2       RPL23P2 0.01373419</span>
<span class="co">#&gt; 2                 BACH1 ENSG00000156273     BACH1    AF129075.2 0.61134838</span>
<span class="co">#&gt; 3                 GABPA ENSG00000154727     GABPA    AP000255.1 0.00000000</span>
<span class="co">#&gt; 4                 GABPA ENSG00000154727     GABPA         BRWD1 0.67481637</span>
<span class="co">#&gt; 5                 GABPA ENSG00000154727     GABPA    AP001625.2 0.61368294</span>
<span class="co">#&gt; 6                PKNOX1 ENSG00000160199    PKNOX1        H2AZP1 0.27934091</span>
<span class="co">#&gt;    pval_rna.tf pval_met:rna.tf estimate_met estimate_rna.tf estimate_met:rna.tf</span>
<span class="co">#&gt; 1 8.927789e-02      0.01263826   -94.880189      -1.8366801           4.5153185</span>
<span class="co">#&gt; 2 9.635728e-03      0.67327726    -9.775541       1.5154069           0.4727370</span>
<span class="co">#&gt; 3 2.341924e-08      0.00000000   258.708769       1.6970768         -15.4781016</span>
<span class="co">#&gt; 4 3.783765e-02      0.62747811     6.141972       1.6780249          -0.4139619</span>
<span class="co">#&gt; 5 6.170294e-01      0.57380835  -167.791692      -9.3709854          10.9969876</span>
<span class="co">#&gt; 6 8.209816e-01      0.28503982   -91.586347      -0.9169839           5.6172434</span>
<span class="co">#&gt;     Model.interaction met.q4_minus_q1 quant_pval_metGrp quant_pval_rna.tf</span>
<span class="co">#&gt; 1 Robust Linear Model      0.18256876      3.953828e-05      5.958024e-03</span>
<span class="co">#&gt; 2 Robust Linear Model      0.31037921      7.437666e-01      6.570509e-04</span>
<span class="co">#&gt; 3 Robust Linear Model      0.08016092      2.208774e-03      5.803942e-01</span>
<span class="co">#&gt; 4 Robust Linear Model      0.04063833      3.823862e-01      7.142112e-08</span>
<span class="co">#&gt; 5 Robust Linear Model      0.01417579      5.619040e-01      1.305771e-03</span>
<span class="co">#&gt; 6 Robust Linear Model      0.14299567      1.524877e-03      3.270442e-01</span>
<span class="co">#&gt;   quant_pval_metGrp:rna.tf quant_estimate_metGrp quant_estimate_rna.tf</span>
<span class="co">#&gt; 1             3.768097e-05            -83.041219            -2.0395999</span>
<span class="co">#&gt; 2             7.945988e-01             -3.533136             1.3437155</span>
<span class="co">#&gt; 3             2.305241e-03            -30.858474             0.2627746</span>
<span class="co">#&gt; 4             4.999473e-01             -3.386762             1.3876278</span>
<span class="co">#&gt; 5             5.234533e-01             -5.525037             1.2125286</span>
<span class="co">#&gt; 6             1.727783e-03           -182.019620             1.8203862</span>
<span class="co">#&gt;   quant_estimate_metGrp:rna.tf      Model.quantile Wilcoxon_pval_tf_q4_vs_q1</span>
<span class="co">#&gt; 1                    3.8764266 Robust Linear Model                 0.5787417</span>
<span class="co">#&gt; 2                    0.1642009 Robust Linear Model                 0.4812509</span>
<span class="co">#&gt; 3                    1.8528803 Robust Linear Model                 0.9117972</span>
<span class="co">#&gt; 4                    0.1522536 Robust Linear Model                 0.7959363</span>
<span class="co">#&gt; 5                    0.3594288 Robust Linear Model                 0.2798610</span>
<span class="co">#&gt; 6                   11.0029788 Robust Linear Model                 0.5288489</span>
<span class="co">#&gt;   % 0 target genes (All samples) % of 0 target genes (Q1 and Q4)</span>
<span class="co">#&gt; 1                         2.63 %                             0 %</span>
<span class="co">#&gt; 2                         5.26 %                             5 %</span>
<span class="co">#&gt; 3                        21.05 %                            20 %</span>
<span class="co">#&gt; 4                            0 %                             0 %</span>
<span class="co">#&gt; 5                         7.89 %                             5 %</span>
<span class="co">#&gt; 6                        10.53 %                            10 %</span>
<span class="co">#&gt;   Max_interaction_pval  fdr_met   fdr_rna.tf fdr_met:rna.tf quant_fdr_metGrp</span>
<span class="co">#&gt; 1          0.012638262 0.576836 7.055858e-01      0.4967681      0.002807218</span>
<span class="co">#&gt; 2          0.794598832 1.000000 2.665885e-01      0.9993441      0.999999995</span>
<span class="co">#&gt; 3          0.002305241 0.000000 1.943797e-06      0.0000000      0.039205740</span>
<span class="co">#&gt; 4          0.627478107 1.000000 6.281050e-01      0.9993441      0.999999995</span>
<span class="co">#&gt; 5          0.573808352 1.000000 1.000000e+00      0.9993441      0.999999995</span>
<span class="co">#&gt; 6          0.285039821 1.000000 1.000000e+00      0.9993441      0.039205740</span>
<span class="co">#&gt;   quant_fdr_rna.tf quant_fdr_metGrp:rna.tf Wilcoxon_fdr_tf_q4_vs_q1</span>
<span class="co">#&gt; 1     8.016700e-02             0.002675349                0.7885356</span>
<span class="co">#&gt; 2     1.598824e-02             0.999999995                0.7388219</span>
<span class="co">#&gt; 3     9.999978e-01             0.040918021                0.9376028</span>
<span class="co">#&gt; 4     5.213742e-06             0.999999995                0.9119969</span>
<span class="co">#&gt; 5     2.383031e-02             0.999999995                0.5546336</span>
<span class="co">#&gt; 6     9.947595e-01             0.040918021                0.7685937</span></pre></div>
</div>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h1>
<div class="sourceCode" id="cb6"><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
<span class="co">#&gt; R version 4.0.2 (2020-06-22)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Catalina 10.15.6</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats4    parallel  stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] BSgenome.Hsapiens.UCSC.hg19_1.4.3 BSgenome_1.56.0                  </span>
<span class="co">#&gt;  [3] rtracklayer_1.48.0                Biostrings_2.56.0                </span>
<span class="co">#&gt;  [5] XVector_0.28.0                    GenomicRanges_1.40.0             </span>
<span class="co">#&gt;  [7] GenomeInfoDb_1.24.2               IRanges_2.22.2                   </span>
<span class="co">#&gt;  [9] S4Vectors_0.26.1                  sesameData_1.6.0                 </span>
<span class="co">#&gt; [11] ExperimentHub_1.14.2              AnnotationHub_2.20.2             </span>
<span class="co">#&gt; [13] BiocFileCache_1.12.1              dbplyr_1.4.4                     </span>
<span class="co">#&gt; [15] BiocGenerics_0.34.0               MethReg_0.1.0                    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] colorspace_1.4-1              ggsignif_0.6.0               </span>
<span class="co">#&gt;   [3] ellipsis_0.3.1                rio_0.5.16                   </span>
<span class="co">#&gt;   [5] ggpubr_0.4.0                  bit64_4.0.2                  </span>
<span class="co">#&gt;   [7] interactiveDisplayBase_1.26.3 AnnotationDbi_1.50.3         </span>
<span class="co">#&gt;   [9] codetools_0.2-16              R.methodsS3_1.8.0            </span>
<span class="co">#&gt;  [11] doParallel_1.0.15             pscl_1.5.5                   </span>
<span class="co">#&gt;  [13] knitr_1.29                    Rsamtools_2.4.0              </span>
<span class="co">#&gt;  [15] seqLogo_1.54.3                annotate_1.66.0              </span>
<span class="co">#&gt;  [17] broom_0.7.0                   GO.db_3.11.4                 </span>
<span class="co">#&gt;  [19] png_0.1-7                     R.oo_1.23.0                  </span>
<span class="co">#&gt;  [21] sfsmisc_1.1-7                 shiny_1.5.0                  </span>
<span class="co">#&gt;  [23] BiocManager_1.30.10           readr_1.3.1                  </span>
<span class="co">#&gt;  [25] compiler_4.0.2                httr_1.4.2                   </span>
<span class="co">#&gt;  [27] backports_1.1.9               assertthat_0.2.1             </span>
<span class="co">#&gt;  [29] Matrix_1.2-18                 fastmap_1.0.1                </span>
<span class="co">#&gt;  [31] later_1.1.0.1                 htmltools_0.5.0              </span>
<span class="co">#&gt;  [33] prettyunits_1.1.1             tools_4.0.2                  </span>
<span class="co">#&gt;  [35] gtable_0.3.0                  glue_1.4.1                   </span>
<span class="co">#&gt;  [37] TFMPvalue_0.0.8               GenomeInfoDbData_1.2.3       </span>
<span class="co">#&gt;  [39] reshape2_1.4.4                dplyr_1.0.2                  </span>
<span class="co">#&gt;  [41] rappdirs_0.3.1                Rcpp_1.0.5                   </span>
<span class="co">#&gt;  [43] carData_3.0-4                 Biobase_2.48.0               </span>
<span class="co">#&gt;  [45] cellranger_1.1.0              vctrs_0.3.2                  </span>
<span class="co">#&gt;  [47] iterators_1.0.12              xfun_0.16                    </span>
<span class="co">#&gt;  [49] CNEr_1.24.0                   stringr_1.4.0                </span>
<span class="co">#&gt;  [51] openxlsx_4.1.5                mime_0.9                     </span>
<span class="co">#&gt;  [53] lifecycle_0.2.0               poweRlaw_0.70.6              </span>
<span class="co">#&gt;  [55] gtools_3.8.2                  rstatix_0.6.0                </span>
<span class="co">#&gt;  [57] XML_3.99-0.5                  zlibbioc_1.34.0              </span>
<span class="co">#&gt;  [59] MASS_7.3-52                   scales_1.1.1                 </span>
<span class="co">#&gt;  [61] hms_0.5.3                     promises_1.1.1               </span>
<span class="co">#&gt;  [63] SummarizedExperiment_1.18.2   JASPAR2020_0.99.10           </span>
<span class="co">#&gt;  [65] yaml_2.2.1                    curl_4.3                     </span>
<span class="co">#&gt;  [67] memoise_1.1.0                 ggplot2_3.3.2                </span>
<span class="co">#&gt;  [69] stringi_1.4.6                 RSQLite_2.2.0                </span>
<span class="co">#&gt;  [71] BiocVersion_3.11.1            foreach_1.5.0                </span>
<span class="co">#&gt;  [73] caTools_1.18.0                zip_2.1.0                    </span>
<span class="co">#&gt;  [75] BiocParallel_1.22.0           rlang_0.4.7                  </span>
<span class="co">#&gt;  [77] pkgconfig_2.0.3               matrixStats_0.56.0           </span>
<span class="co">#&gt;  [79] bitops_1.0-6                  pracma_2.2.9                 </span>
<span class="co">#&gt;  [81] evaluate_0.14                 lattice_0.20-41              </span>
<span class="co">#&gt;  [83] purrr_0.3.4                   GenomicAlignments_1.24.0     </span>
<span class="co">#&gt;  [85] bit_4.0.4                     tidyselect_1.1.0             </span>
<span class="co">#&gt;  [87] plyr_1.8.6                    magrittr_1.5                 </span>
<span class="co">#&gt;  [89] R6_2.4.1                      generics_0.0.2               </span>
<span class="co">#&gt;  [91] DelayedArray_0.14.1           DBI_1.1.0                    </span>
<span class="co">#&gt;  [93] pillar_1.4.6                  haven_2.3.1                  </span>
<span class="co">#&gt;  [95] foreign_0.8-80                KEGGREST_1.28.0              </span>
<span class="co">#&gt;  [97] abind_1.4-5                   RCurl_1.98-1.2               </span>
<span class="co">#&gt;  [99] tibble_3.0.3                  crayon_1.3.4                 </span>
<span class="co">#&gt; [101] car_3.0-9                     rmarkdown_2.3                </span>
<span class="co">#&gt; [103] progress_1.2.2                TFBSTools_1.26.0             </span>
<span class="co">#&gt; [105] grid_4.0.2                    readxl_1.3.1                 </span>
<span class="co">#&gt; [107] data.table_1.13.0             blob_1.2.1                   </span>
<span class="co">#&gt; [109] forcats_0.5.0                 digest_0.6.25                </span>
<span class="co">#&gt; [111] xtable_1.8-4                  tidyr_1.1.1                  </span>
<span class="co">#&gt; [113] httpuv_1.5.4                  R.utils_2.9.2                </span>
<span class="co">#&gt; [115] munsell_0.5.0                 DirichletMultinomial_1.30.0  </span>
<span class="co">#&gt; [117] motifmatchr_1.10.0</span></pre></div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></small></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing MethReg</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Tiago Silva <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-1343-6850" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Lily Wang <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://codecov.io/gl/tiagochst/methtf"><img src="https://codecov.io/gl/tiagochst/methtf/branch/%5Cx6d6173746572/graph/badge.svg?token=NESBYPVF64" alt="codecov"></a></li>
<li><a href=""><img src="https://img.shields.io/badge/license-GPL%20(%3E%3D%202)-blue" alt="license"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Tiago Silva, Lily Wang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
