<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>coMethTF: estimating regulatory potential of DNA methylation changes in epigenome-wide association studies • coMethTF</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="coMethTF: estimating regulatory potential of DNA methylation changes in epigenome-wide association studies">
<meta property="og:description" content="coMethTF">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coMethTF</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/coMethTF.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>coMethTF: estimating regulatory potential of DNA methylation changes in epigenome-wide association studies</h1>
                        <h4 class="author">Tiago Chedraoui Silva<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
                        <h4 class="author">Lily Wang<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>
</h4>
            
      
      
      <div class="hidden name"><code>coMethTF.Rmd</code></div>

    </div>

    
    
<style> body {text-align: justify} </style>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>1. Introduction</h1>
<p>Transcription factors (TFs) are proteins that facilitate the transcription of DNA into RNA. A number of recent studies have observed that the binding of TFs onto DNA can be affected by DNA methylation, and in turn, DNA methylation can also be added or removed by proteins associated with transcription factors <span class="citation">(Bonder et al. 2017; Banovich et al. 2014; Zhu, Wang, and Qian 2016)</span>.</p>
<p>To provide functional annotations for differentially methylated regions (DMRs) and differentially methylated CpG sites (DMS),<code>coMethTF</code> performs integrative analyses using matched DNA methylation and gene expression along with Transcription Factor Binding Sites (TFBS) data. coMethTF evaluates, prioritizes and annotates DNA methylation regions (or sites) with high regulatory potential that works synergistically with TFs to regulate target gene expressions, without any additional ChIP-seq data.</p>
<p>The results from <code>coMethTF</code> can be used to generate testable hypothesis on the synergistic collaboration of DNA methylation changes and TFs in gene regulation. <code>coMethTF</code> can be used either to evaluate regulatory potentials of candidate regions or to search for methylation coupled TF regulatory processes in the entire genome.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>2. Installation</h1>
<p>You can install the development version of the package from GitHub using the following code:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"TransBioInfoLab/coMethTF"</span>)</pre></body></html></div>
<p>After the package is successfully installed you can load its functions with:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">coMethTF</span>)</pre></body></html></div>
</div>
<div id="dataset" class="section level1">
<h1 class="hasAnchor">
<a href="#dataset" class="anchor"></a>3. Dataset</h1>
<p>The input of <code>coMethTF</code> are DNA methylation and gene expression datasets measured on the same set of samples. Using additional information from the JASPAR database, we will then create a data frame with information on the triplet: DMR (or CpG), TF and target gene.</p>
<p>For illustration, we will use chromosome 21 data from 38 TCGA-COAD samples.</p>
<div id="dna-methylation-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#dna-methylation-dataset" class="anchor"></a>3.1 DNA methylation dataset</h2>
<p>The DNA methylation dataset is a matrix with methylation beta or M-values. If there are potential confounding factors (e.g. batch effect, age, sex) in the dataset, this matrix can also contain residuals from fitting linear regression instead (see <strong>Section 5</strong> below).</p>
<p>The samples are in the columns and regions/probes are in the rows.</p>
<div id="analysis-for-individual-cpgs-data" class="section level3">
<h3 class="hasAnchor">
<a href="#analysis-for-individual-cpgs-data" class="anchor"></a>3.1.1 Analysis for individual CpGs data</h3>
<p>The individual CpGs data can be, for example, differentially methylated sites (DMS) obtained in a DNA methylation study.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">dna.met.chr21</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/get.html">get</a></span>(<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"dna.met.chr21"</span>))</pre></body></html></div>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">dna.met.chr21</span>[<span class="fl">1</span>:<span class="fl">5</span>,<span class="fl">1</span>:<span class="fl">5</span>]
<span class="co">#&gt;            TCGA-3L-AA1B-01A TCGA-4N-A93T-01A TCGA-4T-AA8H-01A TCGA-5M-AAT4-01A</span>
<span class="co">#&gt; cg00002080        0.6454046        0.5933725       0.54955509       0.81987982</span>
<span class="co">#&gt; cg00004533        0.9655396        0.9640490       0.96690671       0.95510446</span>
<span class="co">#&gt; cg00009944        0.5437705        0.2803064       0.42918909       0.60734630</span>
<span class="co">#&gt; cg00025591        0.4021317        0.7953653       0.41816364       0.33241304</span>
<span class="co">#&gt; cg00026030        0.1114705        0.1012902       0.06834467       0.08594876</span>
<span class="co">#&gt;            TCGA-5M-AAT5-01A</span>
<span class="co">#&gt; cg00002080       0.79171160</span>
<span class="co">#&gt; cg00004533       0.96061252</span>
<span class="co">#&gt; cg00009944       0.47555585</span>
<span class="co">#&gt; cg00025591       0.67251468</span>
<span class="co">#&gt; cg00026030       0.06715677</span></pre></body></html></div>
<p>We will first create a SummarizedExperiment object with the function <code>make_se_from_dnam_probes</code>. This funtion will use the Sesame package to map the array probes into genomic regions.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">dna.met.chr21.se</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/make_se_from_dnam_probes.html">make_se_from_dnam_probes</a></span>(
  <span class="kw">dnam</span> <span class="kw">=</span> <span class="no">dna.met.chr21</span>,
  <span class="kw">genome</span> <span class="kw">=</span> <span class="st">"hg38"</span>,
  <span class="kw">arrayType</span> <span class="kw">=</span> <span class="st">"450k"</span>
)
<span class="co">#&gt; o Creating a SummarizedExperiment from DNA methylation input</span>
<span class="co">#&gt; oo Fetching probes metadata</span>
<span class="co">#&gt; oo Removing masked probes</span>
<span class="co">#&gt; oo Preparing SummarizedExperiment object</span></pre></body></html></div>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">dna.met.chr21.se</span>
<span class="co">#&gt; class: RangedSummarizedExperiment </span>
<span class="co">#&gt; dim: 2918 38 </span>
<span class="co">#&gt; metadata(0):</span>
<span class="co">#&gt; assays(1): ''</span>
<span class="co">#&gt; rownames(2918): chr21:10450634-10450635 chr21:10520974-10520975 ...</span>
<span class="co">#&gt;   chr21:46670216-46670217 chr21:46670596-46670597</span>
<span class="co">#&gt; rowData names(53): address_A address_B ... MASK_general probeID</span>
<span class="co">#&gt; colnames(38): TCGA-3L-AA1B-01A TCGA-4N-A93T-01A ... TCGA-A6-5656-01B</span>
<span class="co">#&gt;   TCGA-A6-5657-01A</span>
<span class="co">#&gt; colData names(1): samples</span>
<span class="kw pkg">SummarizedExperiment</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">rowRanges</a></span>(<span class="no">dna.met.chr21.se</span>)[<span class="fl">1</span>:<span class="fl">4</span>,<span class="fl">1</span>:<span class="fl">4</span>]
<span class="co">#&gt; GRanges object with 4 ranges and 4 metadata columns:</span>
<span class="co">#&gt;                           seqnames            ranges strand | address_A</span>
<span class="co">#&gt;                              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;</span>
<span class="co">#&gt;   chr21:10450634-10450635    chr21 10450634-10450635      + |  74716393</span>
<span class="co">#&gt;   chr21:10520974-10520975    chr21 10520974-10520975      + |  29756401</span>
<span class="co">#&gt;   chr21:10521044-10521045    chr21 10521044-10521045      + |  15617483</span>
<span class="co">#&gt;   chr21:10521122-10521123    chr21 10521122-10521123      + |  33810384</span>
<span class="co">#&gt;                           address_B     channel  designType</span>
<span class="co">#&gt;                           &lt;integer&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="co">#&gt;   chr21:10450634-10450635      &lt;NA&gt;        Both          II</span>
<span class="co">#&gt;   chr21:10520974-10520975  20622400         Red           I</span>
<span class="co">#&gt;   chr21:10521044-10521045      &lt;NA&gt;        Both          II</span>
<span class="co">#&gt;   chr21:10521122-10521123  37781360         Grn           I</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 26 sequences from an unspecified genome; no seqlengths</span></pre></body></html></div>
</div>
<div id="dmrs-data" class="section level3">
<h3 class="hasAnchor">
<a href="#dmrs-data" class="anchor"></a>3.1.2 DMRs data</h3>
<p>Differentially Methylated Regions (DMRs) associated with phenotypes such as tumor stage can be obtained from R packages such as , ,  and many others. The methylation levels in multiple CpGs within the DMRs need to be summarized (e.g. using medians), then the analysis for DMR will proceed in the same way as those for CpGs. See <strong>Section 6</strong> “Identifying co-methylated clusters and summarizing regions using medians for DMRs” below for more details.</p>
</div>
</div>
<div id="gene-expression-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-expression-dataset" class="anchor"></a>3.2 Gene expression dataset</h2>
<p>The gene expression dataset is a matrix with log2 transformed and normalized gene expression values. If there are potential confounding factors (e.g. batch effect, age, sex) in the dataset, this matrix can also contain residuals from linear regression instead (see <strong>Section 5</strong> below). The samples are in the columns and the genes are in the rows.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">gene.exp.chr21</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/get.html">get</a></span>(<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"gene.exp.chr21"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"coMethTF"</span>))
<span class="no">gene.exp.chr21</span>[<span class="fl">1</span>:<span class="fl">5</span>,<span class="fl">1</span>:<span class="fl">5</span>]
<span class="co">#&gt;                 TCGA-3L-AA1B-01A TCGA-4N-A93T-01A TCGA-4T-AA8H-01A</span>
<span class="co">#&gt; ENSG00000141956         14.64438         14.65342         14.09232</span>
<span class="co">#&gt; ENSG00000141959         19.33519         20.03720         19.76128</span>
<span class="co">#&gt; ENSG00000142149         17.27832         16.02392         18.16079</span>
<span class="co">#&gt; ENSG00000142156         20.38689         18.83080         18.02720</span>
<span class="co">#&gt; ENSG00000142166         17.89172         18.06625         18.47187</span>
<span class="co">#&gt;                 TCGA-5M-AAT4-01A TCGA-5M-AAT5-01A</span>
<span class="co">#&gt; ENSG00000141956         14.60680         14.58640</span>
<span class="co">#&gt; ENSG00000141959         19.57854         18.27442</span>
<span class="co">#&gt; ENSG00000142149         15.84463         14.79654</span>
<span class="co">#&gt; ENSG00000142156         18.91380         18.71926</span>
<span class="co">#&gt; ENSG00000142166         17.40467         16.71412</span></pre></body></html></div>
<p>We will also create a SummarizedExperiment object for the gene expression data. This object will contain the genomic information for each gene.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">gene.exp.chr21.se</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/make_se_from_gene_matrix.html">make_se_from_gene_matrix</a></span>(<span class="no">gene.exp.chr21</span>)
<span class="co">#&gt; o Creating a SummarizedExperiment from gene expression input</span>
<span class="no">gene.exp.chr21.se</span>
<span class="co">#&gt; class: RangedSummarizedExperiment </span>
<span class="co">#&gt; dim: 752 38 </span>
<span class="co">#&gt; metadata(0):</span>
<span class="co">#&gt; assays(1): ''</span>
<span class="co">#&gt; rownames(752): ENSG00000141956 ENSG00000141959 ... ENSG00000281420</span>
<span class="co">#&gt;   ENSG00000281903</span>
<span class="co">#&gt; rowData names(2): ensembl_gene_id external_gene_name</span>
<span class="co">#&gt; colnames(38): TCGA-3L-AA1B-01A TCGA-4N-A93T-01A ... TCGA-A6-5656-01B</span>
<span class="co">#&gt;   TCGA-A6-5657-01A</span>
<span class="co">#&gt; colData names(1): samples</span>
<span class="kw pkg">SummarizedExperiment</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">rowRanges</a></span>(<span class="no">gene.exp.chr21.se</span>)[<span class="fl">1</span>:<span class="fl">5</span>,]
<span class="co">#&gt; GRanges object with 5 ranges and 2 metadata columns:</span>
<span class="co">#&gt;                   seqnames            ranges strand | ensembl_gene_id</span>
<span class="co">#&gt;                      &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt;</span>
<span class="co">#&gt;   ENSG00000141956    chr21 41798225-41879482      - | ENSG00000141956</span>
<span class="co">#&gt;   ENSG00000141959    chr21 44300051-44327376      + | ENSG00000141959</span>
<span class="co">#&gt;   ENSG00000142149    chr21 31873020-32044633      + | ENSG00000142149</span>
<span class="co">#&gt;   ENSG00000142156    chr21 45981769-46005050      + | ENSG00000142156</span>
<span class="co">#&gt;   ENSG00000142166    chr21 33324429-33359864      + | ENSG00000142166</span>
<span class="co">#&gt;                   external_gene_name</span>
<span class="co">#&gt;                          &lt;character&gt;</span>
<span class="co">#&gt;   ENSG00000141956             PRDM15</span>
<span class="co">#&gt;   ENSG00000141959               PFKL</span>
<span class="co">#&gt;   ENSG00000142149               HUNK</span>
<span class="co">#&gt;   ENSG00000142156             COL6A1</span>
<span class="co">#&gt;   ENSG00000142166             IFNAR1</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 24 sequences from an unspecified genome; no seqlengths</span></pre></body></html></div>
</div>
<div id="creating-triplet-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-triplet-dataset" class="anchor"></a>3.3 Creating triplet dataset</h2>
<div id="mapping-dmr-region-or-cpg-to-the-target-gene" class="section level3">
<h3 class="hasAnchor">
<a href="#mapping-dmr-region-or-cpg-to-the-target-gene" class="anchor"></a>Mapping DMR region (or CpG) to the target gene</h3>
<p>The function <code>get_region_target_gene</code> provides two methods to link a DMR to a target gene, by (1) mapping the region to the closest gene (<code>method = "closest.gene"</code>) or (2) mapping the region to all the genes within a window (default size = 500Kbp around the region, i.e. +- 250K bp from start or end of the region) (<code>method = "window"</code>). or (3) mapping the region to a specific number of genes upstream down/upstream of the region (<code>method = "nearest.genes"</code>)</p>
<p>For promoter analysis, we recommend setting <code>method = "closest.gene"</code>. For analysis of distal regions, we recommend setting either <code>method = "window"</code> or <code>method = "nearest.genes"</code>. Note that the distal analysis will be more time and resource consuming.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="co"># Get CpG information</span>
<span class="no">probes.gr</span> <span class="kw">&lt;-</span> <span class="kw pkg">SummarizedExperiment</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">rowRanges</a></span>(<span class="no">dna.met.chr21.se</span>)
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">probes.gr</span>)
<span class="co">#&gt; [1] "GRanges"</span>
<span class="co">#&gt; attr(,"package")</span>
<span class="co">#&gt; [1] "GenomicRanges"</span></pre></body></html></div>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="co"># Map probes to nearest gene</span>
<span class="no">region.target.closest</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/get_region_target_gene.html">get_region_target_gene</a></span>(
    <span class="kw">regions.gr</span> <span class="kw">=</span> <span class="no">probes.gr</span>,
    <span class="kw">genome</span> <span class="kw">=</span> <span class="st">"hg38"</span>,
    <span class="kw">method</span> <span class="kw">=</span> <span class="st">"closest.gene"</span>
)
<span class="co">#&gt; Mapping regions to the closest gene</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">region.target.closest</span>)
<span class="co">#&gt; [1] 988   3</span>
<span class="no">region.target.closest</span>
<span class="co">#&gt; # A tibble: 988 x 3</span>
<span class="co">#&gt;    regionID                target_gene_name target         </span>
<span class="co">#&gt;    &lt;chr&gt;                   &lt;chr&gt;            &lt;chr&gt;          </span>
<span class="co">#&gt;  1 chr21:10520974-10520975 TPTE             ENSG00000274391</span>
<span class="co">#&gt;  2 chr21:10521044-10521045 TPTE             ENSG00000274391</span>
<span class="co">#&gt;  3 chr21:10521122-10521123 TPTE             ENSG00000274391</span>
<span class="co">#&gt;  4 chr21:10521553-10521554 TPTE             ENSG00000274391</span>
<span class="co">#&gt;  5 chr21:14063939-14063940 ANKRD20A18P      ENSG00000249493</span>
<span class="co">#&gt;  6 chr21:14070786-14070787 RNA5SP488        ENSG00000201812</span>
<span class="co">#&gt;  7 chr21:14071916-14071917 RNA5SP488        ENSG00000201812</span>
<span class="co">#&gt;  8 chr21:14215491-14215492 RBM11            ENSG00000185272</span>
<span class="co">#&gt;  9 chr21:14216042-14216043 RBM11            ENSG00000185272</span>
<span class="co">#&gt; 10 chr21:14216051-14216052 RBM11            ENSG00000185272</span>
<span class="co">#&gt; # … with 978 more rows</span></pre></body></html></div>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="co"># Map probes to genes within 500kb window</span>
<span class="no">region.target.window</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/get_region_target_gene.html">get_region_target_gene</a></span>(
    <span class="kw">regions.gr</span> <span class="kw">=</span> <span class="no">probes.gr</span>,
    <span class="kw">genome</span> <span class="kw">=</span> <span class="st">"hg38"</span>,
    <span class="kw">method</span> <span class="kw">=</span> <span class="st">"window"</span>,
     <span class="kw">window.size</span> <span class="kw">=</span> <span class="fl">500</span> * <span class="fl">10</span>^<span class="fl">3</span>
)
<span class="co">#&gt; Removing regions overlapping promoter regions</span>
<span class="co">#&gt; o Get promoter regions for hg38</span>
<span class="co">#&gt; o Remove promoter regions</span>
<span class="co">#&gt; Mapping regions to genes within a window of size: 5e+05 bp</span>

<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">region.target.window</span>)
<span class="co">#&gt; [1] 12916     3</span>
<span class="no">region.target.window</span>
<span class="co">#&gt; # A tibble: 12,916 x 3</span>
<span class="co">#&gt;    regionID                target          target_gene_name</span>
<span class="co">#&gt;    &lt;chr&gt;                   &lt;chr&gt;           &lt;chr&gt;           </span>
<span class="co">#&gt;  1 chr21:10450634-10450635 ENSG00000275945 EIF3FP1         </span>
<span class="co">#&gt;  2 chr21:10450634-10450635 ENSG00000275592 VN1R7P          </span>
<span class="co">#&gt;  3 chr21:10450634-10450635 ENSG00000278106 AF254983.2      </span>
<span class="co">#&gt;  4 chr21:10450634-10450635 ENSG00000187172 BAGE2           </span>
<span class="co">#&gt;  5 chr21:10450634-10450635 ENSG00000273840 AF254983.1      </span>
<span class="co">#&gt;  6 chr21:10450634-10450635 ENSG00000274391 TPTE            </span>
<span class="co">#&gt;  7 chr21:10450634-10450635 ENSG00000278678 CYCSP41         </span>
<span class="co">#&gt;  8 chr21:10616676-10616677 ENSG00000278106 AF254983.2      </span>
<span class="co">#&gt;  9 chr21:10616676-10616677 ENSG00000187172 BAGE2           </span>
<span class="co">#&gt; 10 chr21:10616676-10616677 ENSG00000273840 AF254983.1      </span>
<span class="co">#&gt; # … with 12,906 more rows</span></pre></body></html></div>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="co"># Map probes to 5 genes upstream and 5 downstream</span>
<span class="no">region.target.nearest.genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/get_region_target_gene.html">get_region_target_gene</a></span>(
    <span class="kw">regions.gr</span> <span class="kw">=</span> <span class="no">probes.gr</span>,
    <span class="kw">genome</span> <span class="kw">=</span> <span class="st">"hg38"</span>,
    <span class="kw">method</span> <span class="kw">=</span> <span class="st">"nearest.genes"</span>,
    <span class="kw">num.flanking.genes</span> <span class="kw">=</span> <span class="fl">5</span>
)
<span class="co">#&gt; Removing regions overlapping promoter regions</span>
<span class="co">#&gt; o Get promoter regions for hg38</span>
<span class="co">#&gt; o Remove promoter regions</span>
<span class="co">#&gt; Identifying genes close to the region</span>
<span class="co">#&gt; Identifying 5 genes downstream to the region</span>
<span class="co">#&gt; Identifying 5 genes upstream of the region</span>
<span class="co">#&gt; Identifying gene position for each region</span>

<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">region.target.window</span>)
<span class="co">#&gt; [1] 12916     3</span>
<span class="no">region.target.window</span>
<span class="co">#&gt; # A tibble: 12,916 x 3</span>
<span class="co">#&gt;    regionID                target          target_gene_name</span>
<span class="co">#&gt;    &lt;chr&gt;                   &lt;chr&gt;           &lt;chr&gt;           </span>
<span class="co">#&gt;  1 chr21:10450634-10450635 ENSG00000275945 EIF3FP1         </span>
<span class="co">#&gt;  2 chr21:10450634-10450635 ENSG00000275592 VN1R7P          </span>
<span class="co">#&gt;  3 chr21:10450634-10450635 ENSG00000278106 AF254983.2      </span>
<span class="co">#&gt;  4 chr21:10450634-10450635 ENSG00000187172 BAGE2           </span>
<span class="co">#&gt;  5 chr21:10450634-10450635 ENSG00000273840 AF254983.1      </span>
<span class="co">#&gt;  6 chr21:10450634-10450635 ENSG00000274391 TPTE            </span>
<span class="co">#&gt;  7 chr21:10450634-10450635 ENSG00000278678 CYCSP41         </span>
<span class="co">#&gt;  8 chr21:10616676-10616677 ENSG00000278106 AF254983.2      </span>
<span class="co">#&gt;  9 chr21:10616676-10616677 ENSG00000187172 BAGE2           </span>
<span class="co">#&gt; 10 chr21:10616676-10616677 ENSG00000273840 AF254983.1      </span>
<span class="co">#&gt; # … with 12,906 more rows</span></pre></body></html></div>
</div>
<div id="mapping-region-to-tf" class="section level3">
<h3 class="hasAnchor">
<a href="#mapping-region-to-tf" class="anchor"></a>Mapping region to TF</h3>
<p>JASPAR 2020 is an open-access database of curated, non-redundant transcription factor (TF)-binding profiles <span class="citation">(Baranasic 2020, <span class="citation">@fornes2020jaspar</span>)</span>, which contains more the 500 human TF motifs.</p>
<p>The function <code>get_tf_in_region</code> uses <code>motifmatchr</code> <span class="citation">(Schep 2020)</span> to scan them for occurrences of motifs. The argument <code>window.size</code> will be use to extend the region to scan for the motifs, for example, a <code>window.size</code> of <code>50</code> will add <code>25</code> bp upstream and <code>25</code> bp downstream of the original region.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">region.tf</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/get_tf_in_region.html">get_tf_in_region</a></span>(
    <span class="kw">region</span> <span class="kw">=</span> <span class="no">probes.gr</span>,
    <span class="kw">genome</span> <span class="kw">=</span> <span class="st">"hg38"</span>,
    <span class="kw">window.size</span> <span class="kw">=</span> <span class="fl">50</span>,
    <span class="kw">p.cutoff</span> <span class="kw">=</span> <span class="fl">5e-5</span>
)
<span class="co">#&gt; </span>
<span class="co">#&gt; Evaluating 554 JASPAR Human TF motifs</span>
<span class="co">#&gt; This may take a while...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'S4Vectors'</span>
<span class="co">#&gt; The following objects are masked from 'package:dplyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     first, rename</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     expand.grid</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'IRanges'</span>
<span class="co">#&gt; The following objects are masked from 'package:dplyr':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     collapse, desc, slice</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'Biostrings'</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     strsplit</span>
<span class="co">#&gt; Preparing output</span></pre></body></html></div>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">region.tf</span>)
<span class="co">#&gt;                  regionID TF_external_gene_name              TF</span>
<span class="co">#&gt; 1 chr21:25735671-25735672                 FOXF2 ENSG00000137273</span>
<span class="co">#&gt; 2 chr21:29298095-29298096                 FOXF2 ENSG00000137273</span>
<span class="co">#&gt; 3 chr21:37074284-37074285                 FOXF2 ENSG00000137273</span>
<span class="co">#&gt; 4 chr21:38404512-38404513                 FOXF2 ENSG00000137273</span>
<span class="co">#&gt; 5 chr21:41165367-41165368                 FOXF2 ENSG00000137273</span>
<span class="co">#&gt; 6 chr21:41762207-41762208                 FOXF2 ENSG00000137273</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">region.tf</span>)
<span class="co">#&gt; [1] 8595    3</span></pre></body></html></div>
</div>
<div id="triplet-data-frame-matching-dmr-or-cpgs-with-tfs-and-target-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#triplet-data-frame-matching-dmr-or-cpgs-with-tfs-and-target-genes" class="anchor"></a>Triplet data frame – matching DMR (or CpGs) with TFs and target genes</h3>
<p>The triplet is a data frame with the following columns:</p>
<ul>
<li>
<code>target</code>: gene identifier (obtained from row names of the gene expression matrix),</li>
<li>
<code>regionID</code>: region/cpg identifier (obtained from row names of the DNA methylation matrix)</li>
<li>
<code>TF</code>: gene identifier (obtained from the row names of the gene expression matrix)</li>
</ul>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="co"># Merging the results from the previous sections (region and target, and region and TF)</span>
<span class="no">triplet</span> <span class="kw">&lt;-</span> <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">inner_join</a></span>(<span class="no">region.target.closest</span>, <span class="no">region.tf</span>)</pre></body></html></div>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="co"># Since our data has only chr21 we will only keep TFs on chr21.</span>
<span class="no">triplet</span> <span class="kw">&lt;-</span> <span class="no">triplet</span>[<span class="no">triplet</span>$<span class="no">TF</span> <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">gene.exp.chr21.se</span>),]</pre></body></html></div>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">triplet</span>)
<span class="co">#&gt; # A tibble: 6 x 5</span>
<span class="co">#&gt;   regionID          target_gene_name target       TF_external_gene_n… TF        </span>
<span class="co">#&gt;   &lt;chr&gt;             &lt;chr&gt;            &lt;chr&gt;        &lt;chr&gt;               &lt;chr&gt;     </span>
<span class="co">#&gt; 1 chr21:25562143-2… MIR155HG         ENSG0000023… GABPA               ENSG00000…</span>
<span class="co">#&gt; 2 chr21:25734910-2… ATP5PF           ENSG0000015… ERG                 ENSG00000…</span>
<span class="co">#&gt; 3 chr21:25734910-2… ATP5PF           ENSG0000015… ETS2                ENSG00000…</span>
<span class="co">#&gt; 4 chr21:26846277-2… ADAMTS1          ENSG0000015… PKNOX1              ENSG00000…</span>
<span class="co">#&gt; 5 chr21:28885068-2… N6AMT1           ENSG0000015… PKNOX1              ENSG00000…</span>
<span class="co">#&gt; 6 chr21:28992856-2… LTN1             ENSG0000019… ERG                 ENSG00000…</span></pre></body></html></div>
<p>There may also be multiple rows for a region, when multiple target gene and/or TFs are found close to it.</p>
</div>
</div>
</div>
<div id="evaluating-the-regulatory-potential-of-dmrs-or-cpgs" class="section level1">
<h1 class="hasAnchor">
<a href="#evaluating-the-regulatory-potential-of-dmrs-or-cpgs" class="anchor"></a>4. Evaluating the regulatory potential of DMRs (or CpGs)</h1>
<p>Because TF binding to DNA can be influenced by (or influence) DNA methylation levels nearby <span class="citation">(Yin et al. 2017)</span>, target gene expression can often be influenced by the synergistic effects of both TF and DNA methylation. In other words, TF activities in gene regulation is affected by DNA methylation or vice versa.</p>
<p>Our goal then is to highlight DNA methylation regions (or CpGs) where these synergistic DNAm and TF collaborations occur. We will perform analyses using the 3 datasets described above in Section 3:</p>
<ul>
<li>DNA methylation matrix</li>
<li>Gene expression matrix</li>
<li>The triplet data frame with DNA methylation region, TF, and target gene</li>
</ul>
<div id="analysis-using-model-with-methylation-by-tf-interaction" class="section level2">
<h2 class="hasAnchor">
<a href="#analysis-using-model-with-methylation-by-tf-interaction" class="anchor"></a>4.1 Analysis using model with methylation by TF interaction</h2>
<p>The function <code>interaction_model</code> assess the regulatory impact of DNA methylation on TF regulation of target genes via two approaches:</p>
<p><strong>Model 1</strong> (considering DNAm values as a continuous variable) - we fit a model with TF by DNA methylation interaction using all samples.</p>
<p><span class="math display">\[log_2(RNA target)  \sim log_2(TF) + DNAm + log_2(TF) * DNAm\]</span></p>
<p><strong>Model 2</strong> (considering DNAm values as a binary variable) - we define a binary variable <code>DNAm Group</code> for DNA methylation values (high = 1, low = 0). That is, samples with the highest DNAm levels (top 25 percent) has high = 1, samples with lowest DNAm levels (bottom 25 pecent) has high = 0.</p>
<p>Note that in this implementation, only samples wih DNAm values in the first and last quartiles are considered.</p>
<p><span class="math display">\[log_2(RNA target) \sim log_2(TF) + \text{DNAm Group} + log_2(TF) * \text{DNAm Group}\]</span></p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="no">results.interaction.model</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/interaction_model.html">interaction_model</a></span>(
    <span class="kw">triplet</span> <span class="kw">=</span> <span class="no">triplet</span>,
    <span class="kw">dnam</span> <span class="kw">=</span> <span class="no">dna.met.chr21.se</span>,
    <span class="kw">exp</span> <span class="kw">=</span> <span class="no">gene.exp.chr21.se</span>
)</pre></body></html></div>
<p>The output of <code>interaction_model</code> function will be a data frame with the following variables:</p>
<ul>
<li>
<code>pval_&lt;variable&gt;</code>: p-value for a tested variable (methylation or TF), given the other variables included in the model.</li>
<li>
<code>estimate_&lt;variable&gt;</code>: estimated effect for a variable. If estimate &gt; 0, increasing values of the variable corresponds to increased outcome values (target gene expression). If estimate &lt; 0, increasing values of the variable correspond to decreased target gene expression levels.</li>
</ul>
<p>The following columns are provided for the results of fitting <strong>Model 1</strong> to triplet data:</p>
<ul>
<li>
<code>pval_met</code>: p-value for DNA methylation variable</li>
<li>
<code>pval_rna.tf</code> : p-value for TF expression</li>
<li>
<code>pval_met.rna.tf</code>: : p-value for DNA methylation by TF interaction</li>
<li>
<code>estimate_met</code>: estimate for DNA methylation variable</li>
<li>
<code>estimate_rna.tf</code>: estimate for TF expression</li>
<li>
<code>estimate_met.rna.tf</code>: estimate for DNA methylation by TF interaction</li>
</ul>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">results.interaction.model</span>)
<span class="co">#&gt; [1] 38 25</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">results.interaction.model</span>)
<span class="co">#&gt;  [1] "regionID"                        "target_gene_name"               </span>
<span class="co">#&gt;  [3] "target"                          "TF_external_gene_name"          </span>
<span class="co">#&gt;  [5] "TF"                              "TF_symbol"                      </span>
<span class="co">#&gt;  [7] "target_symbol"                   "pval_met"                       </span>
<span class="co">#&gt;  [9] "pval_rna.tf"                     "pval_met:rna.tf"                </span>
<span class="co">#&gt; [11] "estimate_met"                    "estimate_rna.tf"                </span>
<span class="co">#&gt; [13] "estimate_met:rna.tf"             "Model.interaction"              </span>
<span class="co">#&gt; [15] "met.q4_minus_q1"                 "quant_pval_metGrp"              </span>
<span class="co">#&gt; [17] "quant_pval_rna.tf"               "quant_pval_metGrp:rna.tf"       </span>
<span class="co">#&gt; [19] "quant_estimate_metGrp"           "quant_estimate_rna.tf"          </span>
<span class="co">#&gt; [21] "quant_estimate_metGrp:rna.tf"    "Model.quantile"                 </span>
<span class="co">#&gt; [23] "% 0 target genes (All samples)"  "% of 0 target genes (Q1 and Q4)"</span>
<span class="co">#&gt; [25] "Max_interaction_pval"</span>

<span class="co"># Results for full model (with all samples)</span>
<span class="no">results.interaction.model</span> <span class="kw">%&gt;%</span> <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(
    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">4</span>,<span class="fl">5</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"quant"</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">results.interaction.model</span>),<span class="kw">invert</span> <span class="kw">=</span> <span class="fl">TRUE</span>))
    ) <span class="kw">%&gt;%</span> <span class="no">head</span>
<span class="co">#&gt;                  regionID TF_external_gene_name              TF</span>
<span class="co">#&gt; 1 chr21:25562143-25562144                 GABPA ENSG00000154727</span>
<span class="co">#&gt; 2 chr21:25734910-25734911                   ERG ENSG00000157554</span>
<span class="co">#&gt; 3 chr21:25734910-25734911                  ETS2 ENSG00000157557</span>
<span class="co">#&gt; 4 chr21:26846277-26846278                PKNOX1 ENSG00000160199</span>
<span class="co">#&gt; 5 chr21:28885068-28885069                PKNOX1 ENSG00000160199</span>
<span class="co">#&gt; 6 chr21:28992856-28992857                   ERG ENSG00000157554</span>
<span class="co">#&gt;   target_gene_name          target TF_symbol target_symbol     pval_met</span>
<span class="co">#&gt; 1         MIR155HG ENSG00000234883     GABPA      MIR155HG 0.0081677856</span>
<span class="co">#&gt; 2           ATP5PF ENSG00000154723       ERG        ATP5PF 0.0411445301</span>
<span class="co">#&gt; 3           ATP5PF ENSG00000154723      ETS2        ATP5PF 0.0305598152</span>
<span class="co">#&gt; 4          ADAMTS1 ENSG00000154734    PKNOX1       ADAMTS1 0.6011803496</span>
<span class="co">#&gt; 5           N6AMT1 ENSG00000156239    PKNOX1        N6AMT1 0.7082932153</span>
<span class="co">#&gt; 6             LTN1 ENSG00000198862       ERG          LTN1 0.0007002924</span>
<span class="co">#&gt;   pval_rna.tf pval_met:rna.tf estimate_met estimate_rna.tf estimate_met:rna.tf</span>
<span class="co">#&gt; 1  0.03416955    0.0065915720   -2115.9323     -2.21701395         129.3841033</span>
<span class="co">#&gt; 2  0.18779991    0.0396042166     222.3023      0.45521710         -17.3631376</span>
<span class="co">#&gt; 3  0.02636596    0.0301446441   -1104.9200     -1.73905127          51.8116457</span>
<span class="co">#&gt; 4  0.95310240    0.6478985183     -14.7225     -0.04145077           0.7896601</span>
<span class="co">#&gt; 5  0.77729667    0.7147556592    -114.0424      0.29547948           6.6787547</span>
<span class="co">#&gt; 6  0.02447757    0.0008833081    -305.0305     -1.04512987          20.8024190</span>
<span class="co">#&gt;     Model.interaction met.q4_minus_q1 % 0 target genes (All samples)</span>
<span class="co">#&gt; 1 Robust Linear Model     0.003748681                            0 %</span>
<span class="co">#&gt; 2 Robust Linear Model     0.010247921                            0 %</span>
<span class="co">#&gt; 3 Robust Linear Model     0.010247921                            0 %</span>
<span class="co">#&gt; 4 Robust Linear Model     0.487499687                            0 %</span>
<span class="co">#&gt; 5 Robust Linear Model     0.015129228                            0 %</span>
<span class="co">#&gt; 6 Robust Linear Model     0.040039409                            0 %</span>
<span class="co">#&gt;   % of 0 target genes (Q1 and Q4) Max_interaction_pval</span>
<span class="co">#&gt; 1                             0 %         0.4111206320</span>
<span class="co">#&gt; 2                             0 %         0.0444664514</span>
<span class="co">#&gt; 3                             0 %         0.3975855540</span>
<span class="co">#&gt; 4                             0 %         0.7789040658</span>
<span class="co">#&gt; 5                             0 %         0.7542838153</span>
<span class="co">#&gt; 6                             0 %         0.0008833081</span></pre></body></html></div>
<p>The following columns are provided for the results of fitting <strong>Model 2</strong> (quartile model) to triplet data:</p>
<ul>
<li>
<code>quant_pval_metGrp</code>: p-value for DNA methylation variable</li>
<li>
<code>quant_pval_rna.tf</code> : p-value for TF expression</li>
<li>
<code>quant_pval_metGrp:rna.tf</code>: : p-value for DNA methylation by TF interaction</li>
<li>
<code>quant_estimates_metGrp</code>: estimate for DNA methylation variable</li>
<li>
<code>quant_estimates_rna.tf</code>: estimate for TF expression</li>
<li>
<code>quant_estimates_metGrp:rna.tf</code>: estimate for DNA methylation by TF interaction</li>
</ul>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="co"># Results for quartile model</span>
<span class="no">results.interaction.model</span> <span class="kw">%&gt;%</span> <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(
  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">1</span>,<span class="fl">4</span>,<span class="fl">5</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="st">"quant"</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">results.interaction.model</span>)))
  ) <span class="kw">%&gt;%</span> <span class="no">head</span>
<span class="co">#&gt;                  regionID TF_external_gene_name              TF</span>
<span class="co">#&gt; 1 chr21:25562143-25562144                 GABPA ENSG00000154727</span>
<span class="co">#&gt; 2 chr21:25734910-25734911                   ERG ENSG00000157554</span>
<span class="co">#&gt; 3 chr21:25734910-25734911                  ETS2 ENSG00000157557</span>
<span class="co">#&gt; 4 chr21:26846277-26846278                PKNOX1 ENSG00000160199</span>
<span class="co">#&gt; 5 chr21:28885068-28885069                PKNOX1 ENSG00000160199</span>
<span class="co">#&gt; 6 chr21:28992856-28992857                   ERG ENSG00000157554</span>
<span class="co">#&gt;   quant_pval_metGrp quant_pval_rna.tf quant_pval_metGrp:rna.tf</span>
<span class="co">#&gt; 1      4.553007e-01         0.1118582             4.111206e-01</span>
<span class="co">#&gt; 2      5.102383e-02         0.2112489             4.446645e-02</span>
<span class="co">#&gt; 3      3.786296e-01         0.2438708             3.975856e-01</span>
<span class="co">#&gt; 4      7.475811e-01         0.9994336             7.789041e-01</span>
<span class="co">#&gt; 5      7.590637e-01         0.3130110             7.542838e-01</span>
<span class="co">#&gt; 6      1.695171e-06         0.6334561             1.692317e-06</span>
<span class="co">#&gt;   quant_estimate_metGrp quant_estimate_rna.tf quant_estimate_metGrp:rna.tf</span>
<span class="co">#&gt; 1             -6.521779          0.7448068805                    0.4279239</span>
<span class="co">#&gt; 2             12.057209          0.4120340913                   -0.8690609</span>
<span class="co">#&gt; 3            -12.876617         -0.4733414619                    0.5888877</span>
<span class="co">#&gt; 4             -7.546100         -0.0007817679                    0.4034814</span>
<span class="co">#&gt; 5             -3.410472          0.4557956257                    0.2130744</span>
<span class="co">#&gt; 6            -27.661213         -0.0838915000                    1.9025146</span>
<span class="co">#&gt;        Model.quantile</span>
<span class="co">#&gt; 1 Robust Linear Model</span>
<span class="co">#&gt; 2 Robust Linear Model</span>
<span class="co">#&gt; 3 Robust Linear Model</span>
<span class="co">#&gt; 4 Robust Linear Model</span>
<span class="co">#&gt; 5 Robust Linear Model</span>
<span class="co">#&gt; 6 Robust Linear Model</span></pre></body></html></div>
<p>To select significant DNAm by TF intereaction from both models, you can subset the results using the p-value from each linear model.</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">results.sig</span>)
<span class="co">#&gt;                  regionID target_gene_name          target</span>
<span class="co">#&gt; 1 chr21:25734910-25734911           ATP5PF ENSG00000154723</span>
<span class="co">#&gt; 2 chr21:28992856-28992857             LTN1 ENSG00000198862</span>
<span class="co">#&gt; 3 chr21:32727998-32727999            SYNJ1 ENSG00000159082</span>
<span class="co">#&gt; 4 chr21:32728004-32728005            SYNJ1 ENSG00000159082</span>
<span class="co">#&gt; 5 chr21:36320159-36320160       AP000692.2 ENSG00000273199</span>
<span class="co">#&gt; 6 chr21:36320159-36320160       AP000692.2 ENSG00000273199</span>
<span class="co">#&gt;   TF_external_gene_name              TF TF_symbol target_symbol     pval_met</span>
<span class="co">#&gt; 1                   ERG ENSG00000157554       ERG        ATP5PF 4.114453e-02</span>
<span class="co">#&gt; 2                   ERG ENSG00000157554       ERG          LTN1 7.002924e-04</span>
<span class="co">#&gt; 3                 GABPA ENSG00000154727     GABPA         SYNJ1 1.801255e-03</span>
<span class="co">#&gt; 4                 GABPA ENSG00000154727     GABPA         SYNJ1 4.198808e-02</span>
<span class="co">#&gt; 5                   ERG ENSG00000157554       ERG    AP000692.2 9.042327e-09</span>
<span class="co">#&gt; 6                 GABPA ENSG00000154727     GABPA    AP000692.2 4.440892e-16</span>
<span class="co">#&gt;    pval_rna.tf pval_met:rna.tf estimate_met estimate_rna.tf estimate_met:rna.tf</span>
<span class="co">#&gt; 1 1.877999e-01    3.960422e-02     222.3023      0.45521710           -17.36314</span>
<span class="co">#&gt; 2 2.447757e-02    8.833081e-04    -305.0305     -1.04512987            20.80242</span>
<span class="co">#&gt; 3 9.134597e-01    1.677352e-03    -358.6139     -0.01611787            21.05786</span>
<span class="co">#&gt; 4 3.809482e-01    2.438462e-02    -322.0151     -0.18934706            19.85222</span>
<span class="co">#&gt; 5 1.030086e-07    5.706180e-09    4988.0904      7.87857452          -345.71283</span>
<span class="co">#&gt; 6 0.000000e+00    0.000000e+00    8723.0213     12.71190486          -518.17666</span>
<span class="co">#&gt;     Model.interaction met.q4_minus_q1 quant_pval_metGrp quant_pval_rna.tf</span>
<span class="co">#&gt; 1 Robust Linear Model     0.010247921      5.102383e-02         0.2112489</span>
<span class="co">#&gt; 2 Robust Linear Model     0.040039409      1.695171e-06         0.6334561</span>
<span class="co">#&gt; 3 Robust Linear Model     0.003140598      9.408399e-04         0.2198985</span>
<span class="co">#&gt; 4 Robust Linear Model     0.004731054      3.762372e-04         0.3435961</span>
<span class="co">#&gt; 5 Robust Linear Model     0.006767978      3.306642e-05         0.1443537</span>
<span class="co">#&gt; 6 Robust Linear Model     0.006767978      2.497846e-02         0.9388601</span>
<span class="co">#&gt;   quant_pval_metGrp:rna.tf quant_estimate_metGrp quant_estimate_rna.tf</span>
<span class="co">#&gt; 1             4.446645e-02              12.05721            0.41203409</span>
<span class="co">#&gt; 2             1.692317e-06             -27.66121           -0.08389150</span>
<span class="co">#&gt; 3             6.821033e-04             -12.27496            0.10822193</span>
<span class="co">#&gt; 4             2.167466e-04             -11.51378            0.07895693</span>
<span class="co">#&gt; 5             1.366071e-05              63.28780            0.96022029</span>
<span class="co">#&gt; 6             2.422498e-02              83.30638            0.14086602</span>
<span class="co">#&gt;   quant_estimate_metGrp:rna.tf      Model.quantile</span>
<span class="co">#&gt; 1                   -0.8690609 Robust Linear Model</span>
<span class="co">#&gt; 2                    1.9025146 Robust Linear Model</span>
<span class="co">#&gt; 3                    0.7293916 Robust Linear Model</span>
<span class="co">#&gt; 4                    0.7079154 Robust Linear Model</span>
<span class="co">#&gt; 5                   -4.7630396 Robust Linear Model</span>
<span class="co">#&gt; 6                   -4.9912854 Robust Linear Model</span>
<span class="co">#&gt;   % 0 target genes (All samples) % of 0 target genes (Q1 and Q4)</span>
<span class="co">#&gt; 1                            0 %                             0 %</span>
<span class="co">#&gt; 2                            0 %                             0 %</span>
<span class="co">#&gt; 3                            0 %                             0 %</span>
<span class="co">#&gt; 4                            0 %                             0 %</span>
<span class="co">#&gt; 5                        15.79 %                            20 %</span>
<span class="co">#&gt; 6                        15.79 %                            20 %</span>
<span class="co">#&gt;   Max_interaction_pval</span>
<span class="co">#&gt; 1         4.446645e-02</span>
<span class="co">#&gt; 2         8.833081e-04</span>
<span class="co">#&gt; 3         1.677352e-03</span>
<span class="co">#&gt; 4         2.438462e-02</span>
<span class="co">#&gt; 5         1.366071e-05</span>
<span class="co">#&gt; 6         2.422498e-02</span></pre></body></html></div>
</div>
<div id="stratified-analysis-by-high-and-low-dna-methylation-levels" class="section level2">
<h2 class="hasAnchor">
<a href="#stratified-analysis-by-high-and-low-dna-methylation-levels" class="anchor"></a>4.2 Stratified analysis by high and low DNA methylation levels</h2>
<p>For significant triplets that are identified above, we can further assess how gene regulation by TF changes when DNAm is high or low. To this end, the function <code>stratified_model</code> fits two separate models (see below) to only the samples in highest DNAm levels (top 25 percent), and then to only the samples in lowest DNAm levels (bottom 25 percent), separately.</p>
<p><span class="math display">\[\text{Stratified Model: } log_2(RNA target) \sim log_2(TF)\]</span></p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="no">results.stratified.model</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/stratified_model.html">stratified_model</a></span>(
    <span class="kw">triplet</span> <span class="kw">=</span> <span class="no">triplet</span>,
    <span class="kw">dnam</span> <span class="kw">=</span> <span class="no">dna.met.chr21.se</span>,
    <span class="kw">exp</span> <span class="kw">=</span> <span class="no">gene.exp.chr21.se</span>
)</pre></body></html></div>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="no">results.stratified.model</span> <span class="kw">%&gt;%</span> <span class="no">head</span>
<span class="co">#&gt;                  regionID target_gene_name          target</span>
<span class="co">#&gt; 1 chr21:25562143-25562144         MIR155HG ENSG00000234883</span>
<span class="co">#&gt; 2 chr21:25734910-25734911           ATP5PF ENSG00000154723</span>
<span class="co">#&gt; 3 chr21:25734910-25734911           ATP5PF ENSG00000154723</span>
<span class="co">#&gt; 4 chr21:26846277-26846278          ADAMTS1 ENSG00000154734</span>
<span class="co">#&gt; 5 chr21:28885068-28885069           N6AMT1 ENSG00000156239</span>
<span class="co">#&gt; 6 chr21:28992856-28992857             LTN1 ENSG00000198862</span>
<span class="co">#&gt;   TF_external_gene_name              TF TF_symbol target_symbol</span>
<span class="co">#&gt; 1                 GABPA ENSG00000154727     GABPA      MIR155HG</span>
<span class="co">#&gt; 2                   ERG ENSG00000157554       ERG        ATP5PF</span>
<span class="co">#&gt; 3                  ETS2 ENSG00000157557      ETS2        ATP5PF</span>
<span class="co">#&gt; 4                PKNOX1 ENSG00000160199    PKNOX1       ADAMTS1</span>
<span class="co">#&gt; 5                PKNOX1 ENSG00000160199    PKNOX1        N6AMT1</span>
<span class="co">#&gt; 6                   ERG ENSG00000157554       ERG          LTN1</span>
<span class="co">#&gt;   DNAmlow_pval_rna.tf DNAmlow_estimate_rna.tf DNAmhigh_pval_rna.tf</span>
<span class="co">#&gt; 1           0.1632286              0.74308520         0.0009576798</span>
<span class="co">#&gt; 2           0.2413584              0.41203409         0.0862123233</span>
<span class="co">#&gt; 3           0.2283042             -0.46848956         0.8512898169</span>
<span class="co">#&gt; 4           0.9582855             -0.07594386         0.5802354541</span>
<span class="co">#&gt; 5           0.2797653              0.45658887         0.2696539894</span>
<span class="co">#&gt; 6           0.6517833             -0.07363418         0.0761374174</span>
<span class="co">#&gt;   DNAmhigh_estimate_rna.tf TF.affinity   TF.role</span>
<span class="co">#&gt; 1                1.1717745      M-plus Activator</span>
<span class="co">#&gt; 2               -0.4565995      M-plus Repressor</span>
<span class="co">#&gt; 3                0.1166272     M-minus Repressor</span>
<span class="co">#&gt; 4                0.4017364      M-plus Activator</span>
<span class="co">#&gt; 5                0.6688700      M-plus Activator</span>
<span class="co">#&gt; 6                1.0810969      M-plus Activator</span></pre></body></html></div>
</div>
<div id="visualization-of-data" class="section level2">
<h2 class="hasAnchor">
<a href="#visualization-of-data" class="anchor"></a>4.3 Visualization of data</h2>
<p>The function <code>plot_interaction_model</code> will create figures to visualize the data, in a way that corresponds to the linear model we considered above. It requires the output from the function <code>interaction_model</code> (a dataframe), the DNA methylation matrix and the gene expression matrix as input.</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="no">plots</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/plot_stratified_model.html">plot_stratified_model</a></span>(
    <span class="kw">triplet.results</span> <span class="kw">=</span> <span class="no">results.stratified.model</span>,
    <span class="kw">dnam</span> <span class="kw">=</span> <span class="no">dna.met.chr21.se</span>,
    <span class="kw">exp</span> <span class="kw">=</span> <span class="no">gene.exp.chr21.se</span>
)
<span class="co">#&gt; Warning in rlm.default(x, y, weights, method = method, wt.method = wt.method, :</span>
<span class="co">#&gt; 'rlm' failed to converge in 20 steps</span>

<span class="co">#&gt; Warning in rlm.default(x, y, weights, method = method, wt.method = wt.method, :</span>
<span class="co">#&gt; 'rlm' failed to converge in 20 steps</span>

<span class="co">#&gt; Warning in rlm.default(x, y, weights, method = method, wt.method = wt.method, :</span>
<span class="co">#&gt; 'rlm' failed to converge in 20 steps</span>

<span class="co">#&gt; Warning in rlm.default(x, y, weights, method = method, wt.method = wt.method, :</span>
<span class="co">#&gt; 'rlm' failed to converge in 20 steps</span>

<span class="co">#&gt; Warning in rlm.default(x, y, weights, method = method, wt.method = wt.method, :</span>
<span class="co">#&gt; 'rlm' failed to converge in 20 steps</span>

<span class="co">#&gt; Warning in rlm.default(x, y, weights, method = method, wt.method = wt.method, :</span>
<span class="co">#&gt; 'rlm' failed to converge in 20 steps</span>

<span class="co">#&gt; Warning in rlm.default(x, y, weights, method = method, wt.method = wt.method, :</span>
<span class="co">#&gt; 'rlm' failed to converge in 20 steps</span>

<span class="co">#&gt; Warning in rlm.default(x, y, weights, method = method, wt.method = wt.method, :</span>
<span class="co">#&gt; 'rlm' failed to converge in 20 steps</span>

<span class="co">#&gt; Warning in rlm.default(x, y, weights, method = method, wt.method = wt.method, :</span>
<span class="co">#&gt; 'rlm' failed to converge in 20 steps</span></pre></body></html></div>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="no">plots</span><span class="kw">[[</span><span class="fl">1</span>]]</pre></body></html></div>
<p><img src="coMethTF_files/figure-html/unnamed-chunk-25-1.png" width="1152"></p>
<p>The first row of the figures shows pairwise associations between DNA methylation, TF and target gene expression levels.</p>
<p>The second row of the figures show how much TF activity on target gene expression levels vary by DNA methylation levels. When TF by methylation interaction is significant (Section 4.1), we expect the association between TF and target gene expression vary depending on whether DNA methylation is low or high.</p>
<p>In this example, when DNA methylation is low, more abundant TF corresponds to decreased gene expression (a repressor TF). On the other hand, when DNA methylation level is high, target gene expression is relatively independent of the amount of TF available. One possibility is that DNA methylation might affect TF binding in this case.</p>
<p>If you want to highlight different group of samples, you can also use the metadata argument with a dataframe with rownames as samples (same order as the columns of the DNAm and gene expression), and a column highlighting the groups.</p>
<div class="sourceCode" id="cb26"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"clinical"</span>)
<span class="no">metadata</span> <span class="kw">&lt;-</span> <span class="no">clinical</span>[,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"sample_type"</span>),<span class="kw">drop</span> <span class="kw">=</span> <span class="fl">FALSE</span>]</pre></body></html></div>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="no">plots</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/plot_stratified_model.html">plot_stratified_model</a></span>(
        <span class="kw">triplet.results</span> <span class="kw">=</span> <span class="no">results.stratified.model</span>,
        <span class="kw">dnam</span> <span class="kw">=</span> <span class="no">dna.met.chr21.se</span>,
        <span class="kw">exp</span> <span class="kw">=</span> <span class="no">gene.exp.chr21.se</span>,
        <span class="kw">metadata</span> <span class="kw">=</span> <span class="no">metadata</span>
)</pre></body></html></div>
<div class="sourceCode" id="cb28"><html><body><pre class="r"><span class="no">plots</span><span class="kw">[[</span><span class="fl">1</span>]]</pre></body></html></div>
</div>
</div>
<div id="results-interpretation" class="section level1">
<h1 class="hasAnchor">
<a href="#results-interpretation" class="anchor"></a>5. Results interpretation</h1>
<p>Shown below are some expected results from fitting Models 1 &amp; 2 described in <strong>Section 4.1</strong> above, depending on TF binding preferences. Please note that there can be more possible scenarios than those listed here, therefore, careful evaluation of the statistical models and visualizations of data as described in <strong>Section 4</strong> is needed to gain a good understanding of the multi-omics data.</p>
<div class="figure">
<img src="../../../../Dropbox%20(BBSR)/PanCancer/methTF/vignettes/scenarios.png" alt="Illustration - Possible scenarios"><p class="caption">Illustration - Possible scenarios</p>
</div>
</div>
<div id="identifying-co-methylated-clusters-and-summarizing-regions-using-medians-for-dmrs" class="section level1">
<h1 class="hasAnchor">
<a href="#identifying-co-methylated-clusters-and-summarizing-regions-using-medians-for-dmrs" class="anchor"></a>6. Identifying co-methylated clusters and summarizing regions using medians for DMRs</h1>
<p>Given regions for DMRs, the R package <a href="https://github.com/TransBioInfoLab/coMethDMR">coMethDMR</a> can be used to identify co-methylated regions <span class="citation">(Gomez et al. 2019)</span>. To install it, please use the following command:</p>
<div class="sourceCode" id="cb29"><html><body><pre class="r"><span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="st">"TransBioInfoLab/coMethDMR"</span>)</pre></body></html></div>
<p>We can then summarize methylation levels from multiple CpGs within each region using median.</p>
<div class="sourceCode" id="cb30"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">coMethDMR</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>)</pre></body></html></div>
<p>For illustration, we will use probe clusters in predefined genomic regions from the <code>coMethDMR</code> R package here.</p>
<div class="sourceCode" id="cb31"><html><body><pre class="r"><span class="no">probes.cluster</span> <span class="kw">&lt;-</span> <span class="kw pkg">coMethDMR</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/coMethDMR/man/getPredefinedCluster.html">getPredefinedCluster</a></span>(
  <span class="kw">arrayType</span> <span class="kw">=</span> <span class="st">"450k"</span>,
  <span class="kw">clusterType</span> <span class="kw">=</span> <span class="st">"regions"</span>
)</pre></body></html></div>
<div class="sourceCode" id="cb32"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/message.html">message</a></span>(<span class="st">"Number of clusters: "</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">probes.cluster</span>))
<span class="no">probes.cluster</span> <span class="kw">%&gt;%</span> <span class="no">head</span></pre></body></html></div>
<p>For real data analyis, please replace <code>probes.cluster</code> with probes in DMRs identified in epigenomewide association study. The <code><a href="https://rdrr.io/pkg/coMethDMR/man/GetCpGsInAllRegion.html">coMethDMR::GetCpGsInAllRegion</a></code> function can be used to extract probes within the DMRs.</p>
<p>To find co-methylated regions within the DMRs, first we need to extract probes within the DMRs.</p>
<div class="sourceCode" id="cb33"><html><body><pre class="r"><span class="no">probes.cluster</span> <span class="kw">&lt;-</span> <span class="kw pkg">coMethDMR</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/coMethDMR/man/GetCpGsInAllRegion.html">GetCpGsInAllRegion</a></span>(
    <span class="kw">regionName_char</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"chr21:37432493-37432782"</span>,
                        <span class="st">"chr21:18884682-18885210"</span>,
                        <span class="st">"chr21:27107799-27107982"</span>),
      <span class="kw">arrayType</span> <span class="kw">=</span> <span class="st">"450k"</span>
)
<span class="no">probes.cluster</span></pre></body></html></div>
<p>Next we find co-methylated regions within the DMRs.</p>
<div class="sourceCode" id="cb34"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"dna.met.chr21"</span>)
<span class="co"># just keep the used probes</span>
<span class="no">dna.met.chr21</span> <span class="kw">&lt;-</span> <span class="no">dna.met.chr21</span>[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">dna.met.chr21</span>) <span class="kw">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="no">probes.cluster</span>)),]</pre></body></html></div>
<div class="sourceCode" id="cb35"><html><body><pre class="r"><span class="no">coMeth_ls</span> <span class="kw">&lt;-</span> <span class="kw pkg">coMethDMR</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/coMethDMR/man/CoMethAllRegions.html">CoMethAllRegions</a></span>(
    <span class="kw">dnam</span> <span class="kw">=</span> <span class="no">dna.met.chr21</span>,
    <span class="kw">betaToM</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
    <span class="kw">method</span> <span class="kw">=</span> <span class="st">"spearman"</span>,
    <span class="kw">minPairwiseCorr</span> <span class="kw">=</span> <span class="fl">0.2</span>,
    <span class="kw">CpGs_ls</span> <span class="kw">=</span> <span class="no">probes.cluster</span>,
    <span class="kw">arrayType</span> <span class="kw">=</span> <span class="st">"450k"</span>,
    <span class="kw">returnAllCpGs</span> <span class="kw">=</span> <span class="fl">FALSE</span>,
    <span class="kw">output</span> <span class="kw">=</span> <span class="st">"CpGs"</span>,
    <span class="kw">nCores_int</span> <span class="kw">=</span> <span class="fl">1</span>
)</pre></body></html></div>
<div class="sourceCode" id="cb36"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/message.html">message</a></span>(<span class="st">"Number of co-methylated cluster: "</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="no">coMeth_ls</span>))
<span class="no">coMeth_ls</span> <span class="kw">%&gt;%</span> <span class="no">head</span></pre></body></html></div>
<p>Finally, we compute median methylation of co-methylated regions.</p>
<div class="sourceCode" id="cb37"><html><body><pre class="r"><span class="no">median.df.tumor</span> <span class="kw">&lt;-</span> <span class="kw pkg">plyr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyr/man/ldply.html">ldply</a></span>(
    <span class="no">coMeth_ls</span>[!<span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">coMeth_ls</span>))],
    <span class="kw">function</span>(<span class="no">probes</span>){
        <span class="kw pkg">matrixStats</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowMedians.html">colMedians</a></span>(<span class="no">dna.met.chr21</span>[<span class="no">probes</span>,])
        }, <span class="kw">.progress</span> <span class="kw">=</span> <span class="st">"time"</span>,<span class="kw">.parallel</span> <span class="kw">=</span> <span class="fl">FALSE</span>
)</pre></body></html></div>
<div class="sourceCode" id="cb38"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">median.df.tumor</span>) <span class="kw">&lt;-</span> <span class="no">median.df.tumor</span>$<span class="no">.id</span>
<span class="no">median.df.tumor</span>$<span class="no">.id</span> <span class="kw">&lt;-</span> <span class="kw">NULL</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">median.df.tumor</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">dna.met.chr21</span>)
<span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(<span class="no">median.df.tumor</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">median.df.tumor</span>)</pre></body></html></div>
<p>The analysis described in <strong>Section 3.3</strong> and <strong>Section 4</strong> can then be performed for DMRs using <code>median.df.tumor</code>, in the same way as those for dataset <code>dna.met.chr21</code>.</p>
</div>
<div id="controlling-effects-from-confounding-variables" class="section level1">
<h1 class="hasAnchor">
<a href="#controlling-effects-from-confounding-variables" class="anchor"></a>7. Controlling effects from confounding variables</h1>
<p>Both gene expressions and DNA methylation levels can be affected by age, sex, shifting in cell types, batch effects and other confounding variables.<br>
In this section, we illustrate analysis workflow that reduces confounding effects, by first extracting the residual data with the function <code>get_residuals</code>, before fitting the models discussed above in Section 4.</p>
<p>The <code>get_residuals</code> function will use gene expression (or DNA methylation data) and phenotype data as input. To remove confounding effects in gene expression data, we use the <code>get_residuals</code> function which extract residuals after fitting the following model: <span class="math display">\[features \sim covariate_{1} + covariate_{2} ... + covariate_{N}\]</span></p>
<div class="sourceCode" id="cb39"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"gene.exp.chr21"</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"clinical"</span>)
<span class="no">metadata</span> <span class="kw">&lt;-</span> <span class="no">clinical</span>[,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"sample_type"</span>,<span class="st">"gender"</span>)]

<span class="no">gene.exp.chr21.residuals</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/get_residuals.html">get_residuals</a></span>(<span class="no">gene.exp.chr21</span>[<span class="fl">1</span>:<span class="fl">250</span>,], <span class="no">metadata</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()</pre></body></html></div>
<div class="sourceCode" id="cb40"><html><body><pre class="r"><span class="no">gene.exp.chr21.residuals</span>[<span class="fl">1</span>:<span class="fl">5</span>,<span class="fl">1</span>:<span class="fl">5</span>]</pre></body></html></div>
<div class="sourceCode" id="cb41"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"dna.met.chr21"</span>)
<span class="no">dna.met.chr21</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/map_probes_to_regions.html">map_probes_to_regions</a></span>(<span class="no">dna.met.chr21</span>)
<span class="no">dna.met.chr21.residuals</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/get_residuals.html">get_residuals</a></span>(<span class="no">dna.met.chr21</span>, <span class="no">metadata</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()</pre></body></html></div>
<div class="sourceCode" id="cb42"><html><body><pre class="r"><span class="no">dna.met.chr21.residuals</span>[<span class="fl">1</span>:<span class="fl">5</span>,<span class="fl">1</span>:<span class="fl">5</span>]</pre></body></html></div>
<p>The <code>interaction_model</code> function can then be applied to these residuals data, as the following: <code>rna.target ~ met + rna.tf + rna.tf * met</code>.</p>
<div class="sourceCode" id="cb43"><html><body><pre class="r"><span class="no">results</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/interaction_model.html">interaction_model</a></span>(
    <span class="kw">triplet</span> <span class="kw">=</span> <span class="no">triplet</span>,
    <span class="kw">dnam</span> <span class="kw">=</span> <span class="no">dna.met.chr21.residuals</span>,
    <span class="kw">exp</span> <span class="kw">=</span> <span class="no">gene.exp.chr21.residuals</span>
)</pre></body></html></div>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>8. Session information</h1>
<div class="sourceCode" id="cb44"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
<span class="co">#&gt; R version 4.0.2 (2020-06-22)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Catalina 10.15.6</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats4    parallel  stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] BSgenome.Hsapiens.UCSC.hg38_1.4.3 BSgenome_1.56.0                  </span>
<span class="co">#&gt;  [3] rtracklayer_1.48.0                Biostrings_2.56.0                </span>
<span class="co">#&gt;  [5] XVector_0.28.0                    GenomicRanges_1.40.0             </span>
<span class="co">#&gt;  [7] GenomeInfoDb_1.24.2               IRanges_2.22.2                   </span>
<span class="co">#&gt;  [9] S4Vectors_0.26.1                  coMethTF_0.1.0                   </span>
<span class="co">#&gt; [11] sesameData_1.6.0                  ExperimentHub_1.14.0             </span>
<span class="co">#&gt; [13] AnnotationHub_2.20.0              BiocFileCache_1.12.0             </span>
<span class="co">#&gt; [15] dbplyr_1.4.4                      BiocGenerics_0.34.0              </span>
<span class="co">#&gt; [17] dplyr_1.0.1                      </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] readxl_1.3.1                  backports_1.1.8              </span>
<span class="co">#&gt;   [3] plyr_1.8.6                    splines_4.0.2                </span>
<span class="co">#&gt;   [5] BiocParallel_1.22.0           ggplot2_3.3.2                </span>
<span class="co">#&gt;   [7] TFBSTools_1.26.0              digest_0.6.25                </span>
<span class="co">#&gt;   [9] foreach_1.5.0                 htmltools_0.5.0              </span>
<span class="co">#&gt;  [11] GO.db_3.11.4                  fansi_0.4.1                  </span>
<span class="co">#&gt;  [13] magrittr_1.5                  memoise_1.1.0                </span>
<span class="co">#&gt;  [15] doParallel_1.0.15             sfsmisc_1.1-7                </span>
<span class="co">#&gt;  [17] openxlsx_4.1.5                readr_1.3.1                  </span>
<span class="co">#&gt;  [19] annotate_1.66.0               matrixStats_0.56.0           </span>
<span class="co">#&gt;  [21] R.utils_2.9.2                 JASPAR2020_0.99.10           </span>
<span class="co">#&gt;  [23] pkgdown_1.5.1                 prettyunits_1.1.1            </span>
<span class="co">#&gt;  [25] colorspace_1.4-1              blob_1.2.1                   </span>
<span class="co">#&gt;  [27] rappdirs_0.3.1                haven_2.3.1                  </span>
<span class="co">#&gt;  [29] xfun_0.16                     crayon_1.3.4                 </span>
<span class="co">#&gt;  [31] RCurl_1.98-1.2                TFMPvalue_0.0.8              </span>
<span class="co">#&gt;  [33] iterators_1.0.12              glue_1.4.1                   </span>
<span class="co">#&gt;  [35] gtable_0.3.0                  zlibbioc_1.34.0              </span>
<span class="co">#&gt;  [37] DelayedArray_0.14.1           car_3.0-8                    </span>
<span class="co">#&gt;  [39] abind_1.4-5                   scales_1.1.1                 </span>
<span class="co">#&gt;  [41] pscl_1.5.5                    DBI_1.1.0                    </span>
<span class="co">#&gt;  [43] rstatix_0.6.0                 Rcpp_1.0.5                   </span>
<span class="co">#&gt;  [45] xtable_1.8-4                  progress_1.2.2               </span>
<span class="co">#&gt;  [47] foreign_0.8-80                bit_4.0.3                    </span>
<span class="co">#&gt;  [49] httr_1.4.2                    ellipsis_0.3.1               </span>
<span class="co">#&gt;  [51] farver_2.0.3                  pkgconfig_2.0.3              </span>
<span class="co">#&gt;  [53] XML_3.99-0.5                  R.methodsS3_1.8.0            </span>
<span class="co">#&gt;  [55] utf8_1.1.4                    labeling_0.3                 </span>
<span class="co">#&gt;  [57] tidyselect_1.1.0              rlang_0.4.7                  </span>
<span class="co">#&gt;  [59] reshape2_1.4.4                later_1.1.0.1                </span>
<span class="co">#&gt;  [61] AnnotationDbi_1.50.3          munsell_0.5.0                </span>
<span class="co">#&gt;  [63] BiocVersion_3.11.1            cellranger_1.1.0             </span>
<span class="co">#&gt;  [65] tools_4.0.2                   cli_2.0.2                    </span>
<span class="co">#&gt;  [67] DirichletMultinomial_1.30.0   generics_0.0.2               </span>
<span class="co">#&gt;  [69] RSQLite_2.2.0                 broom_0.7.0                  </span>
<span class="co">#&gt;  [71] evaluate_0.14                 stringr_1.4.0                </span>
<span class="co">#&gt;  [73] fastmap_1.0.1                 yaml_2.2.1                   </span>
<span class="co">#&gt;  [75] knitr_1.29                    bit64_4.0.2                  </span>
<span class="co">#&gt;  [77] fs_1.5.0                      zip_2.0.4                    </span>
<span class="co">#&gt;  [79] caTools_1.18.0                purrr_0.3.4                  </span>
<span class="co">#&gt;  [81] KEGGREST_1.28.0               nlme_3.1-148                 </span>
<span class="co">#&gt;  [83] mime_0.9                      R.oo_1.23.0                  </span>
<span class="co">#&gt;  [85] poweRlaw_0.70.6               pracma_2.2.9                 </span>
<span class="co">#&gt;  [87] compiler_4.0.2                rstudioapi_0.11              </span>
<span class="co">#&gt;  [89] curl_4.3                      png_0.1-7                    </span>
<span class="co">#&gt;  [91] interactiveDisplayBase_1.26.3 ggsignif_0.6.0               </span>
<span class="co">#&gt;  [93] tibble_3.0.3                  stringi_1.4.6                </span>
<span class="co">#&gt;  [95] desc_1.2.0                    forcats_0.5.0                </span>
<span class="co">#&gt;  [97] lattice_0.20-41               CNEr_1.24.0                  </span>
<span class="co">#&gt;  [99] Matrix_1.2-18                 vctrs_0.3.2                  </span>
<span class="co">#&gt; [101] pillar_1.4.6                  lifecycle_0.2.0              </span>
<span class="co">#&gt; [103] BiocManager_1.30.10           cowplot_1.0.0                </span>
<span class="co">#&gt; [105] data.table_1.13.0             bitops_1.0-6                 </span>
<span class="co">#&gt; [107] httpuv_1.5.4                  R6_2.4.1                     </span>
<span class="co">#&gt; [109] promises_1.1.1                gridExtra_2.3                </span>
<span class="co">#&gt; [111] rio_0.5.16                    codetools_0.2-16             </span>
<span class="co">#&gt; [113] MASS_7.3-51.6                 gtools_3.8.2                 </span>
<span class="co">#&gt; [115] assertthat_0.2.1              seqLogo_1.54.3               </span>
<span class="co">#&gt; [117] SummarizedExperiment_1.18.2   rprojroot_1.3-2              </span>
<span class="co">#&gt; [119] GenomicAlignments_1.24.0      Rsamtools_2.4.0              </span>
<span class="co">#&gt; [121] GenomeInfoDbData_1.2.3        mgcv_1.8-31                  </span>
<span class="co">#&gt; [123] hms_0.5.3                     motifmatchr_1.10.0           </span>
<span class="co">#&gt; [125] grid_4.0.2                    prettydoc_0.3.1              </span>
<span class="co">#&gt; [127] tidyr_1.1.1                   rmarkdown_2.3                </span>
<span class="co">#&gt; [129] carData_3.0-4                 ggpubr_0.4.0                 </span>
<span class="co">#&gt; [131] Biobase_2.48.0                shiny_1.5.0</span></pre></body></html></div>
</div>
<div id="bibliography" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#bibliography" class="anchor"></a>9. Bibliography</h1>
<div id="refs" class="references">
<div id="ref-banovich2014methylation">
<p>Banovich, Nicholas E, Xun Lan, Graham McVicker, Bryce Van de Geijn, Jacob F Degner, John D Blischak, Julien Roux, Jonathan K Pritchard, and Yoav Gilad. 2014. “Methylation Qtls Are Associated with Coordinated Changes in Transcription Factor Binding, Histone Modifications, and Gene Expression Levels.” <em>PLoS Genetics</em> 10 (9). Public Library of Science.</p>
</div>
<div id="ref-JASPAR2020">
<p>Baranasic, Damir. 2020. <em>JASPAR2020: Data Package for Jaspar Database (Version 2020)</em>. <a href="http://jaspar.genereg.net/">http://jaspar.genereg.net/</a>.</p>
</div>
<div id="ref-bonder2017disease">
<p>Bonder, Marc Jan, René Luijk, Daria V Zhernakova, Matthijs Moed, Patrick Deelen, Martijn Vermaat, Maarten Van Iterson, et al. 2017. “Disease Variants Alter Transcription Factor Levels and Methylation of Their Binding Sites.” <em>Nature Genetics</em> 49 (1). Nature Publishing Group: 131.</p>
</div>
<div id="ref-fornes2020jaspar">
<p>Fornes, Oriol, Jaime A Castro-Mondragon, Aziz Khan, Robin Van der Lee, Xi Zhang, Phillip A Richmond, Bhavi P Modi, et al. 2020. “JASPAR 2020: Update of the Open-Access Database of Transcription Factor Binding Profiles.” <em>Nucleic Acids Research</em> 48 (D1). Oxford University Press: D87–D92.</p>
</div>
<div id="ref-gomez2019comethdmr">
<p>Gomez, Lissette, Gabriel J Odom, Juan I Young, Eden R Martin, Lizhong Liu, Xi Chen, Anthony J Griswold, Zhen Gao, Lanyu Zhang, and Lily Wang. 2019. “CoMethDMR: Accurate Identification of Co-Methylated and Differentially Methylated Regions in Epigenome-Wide Association Studies with Continuous Phenotypes.” <em>Nucleic Acids Research</em> 47 (17). Oxford University Press: e98–e98.</p>
</div>
<div id="ref-motifmatchr">
<p>Schep, Alicia. 2020. <em>Motifmatchr: Fast Motif Matching in R</em>.</p>
</div>
<div id="ref-yin2017impact">
<p>Yin, Yimeng, Ekaterina Morgunova, Arttu Jolma, Eevi Kaasinen, Biswajyoti Sahu, Syed Khund-Sayeed, Pratyush K Das, et al. 2017. “Impact of Cytosine Methylation on Dna Binding Specificities of Human Transcription Factors.” <em>Science</em> 356 (6337). American Association for the Advancement of Science.</p>
</div>
<div id="ref-zhu2016transcription">
<p>Zhu, Heng, Guohua Wang, and Jiang Qian. 2016. “Transcription Factors as Readers and Effectors of Dna Methylation.” <em>Nature Reviews Genetics</em> 17 (9). Nature Publishing Group: 551–65.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>University of Miami Miller School of Medicine, txs902 at miami.edu<a href="#fnref1" class="footnote-back">↩</a></p></li>
<li id="fn2"><p>University of Miami Miller School of Medicine, lily.wang at gmail.com<a href="#fnref2" class="footnote-back">↩</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Tiago Silva, Lily Wang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
