<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Map DNAm to target genes using distance approaches, and
TF to the DNAm region using JASPAS2020 TFBS. — create_triplet_distance_based • MethReg</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Map DNAm to target genes using distance approaches, and
TF to the DNAm region using JASPAS2020 TFBS. — create_triplet_distance_based" />
<meta property="og:description" content="This function wraps two other functions
get_region_target_gene and get_tf_in_region from the package.
This function will map a region to a target gene using three methods
(mapping to the closest gene,
mapping to any gene within a given window of distance, or mapping to a
fixed number of nearby genes upstream or downstream).
To find TFs binding to the region, JASPAR2020 is used." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MethReg</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/MethReg.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Map DNAm to target genes using distance approaches, and
TF to the DNAm region using JASPAS2020 TFBS.</h1>
    
    <div class="hidden name"><code>create_triplet_distance_based.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function wraps two other functions
<code>get_region_target_gene</code> and <code>get_tf_in_region</code> from the package.
This function will map a region to a target gene using three methods
(mapping to the closest gene,
mapping to any gene within a given window of distance, or mapping to a
fixed number of nearby genes upstream or downstream).
To find TFs binding to the region, JASPAR2020 is used.</p>
    </div>

    <pre class="usage"><span class='fu'>create_triplet_distance_based</span>(
  <span class='no'>region</span>,
  <span class='kw'>genome</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"hg38"</span>, <span class='st'>"hg19"</span>),
  <span class='kw'>target.method</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"genes.promoter.overlap"</span>, <span class='st'>"window"</span>, <span class='st'>"nearby.genes"</span>, <span class='st'>"closest.gene"</span>),
  <span class='kw'>target.window.size</span> <span class='kw'>=</span> <span class='fl'>500</span> * <span class='fl'>10</span>^<span class='fl'>3</span>,
  <span class='kw'>target.num.flanking.genes</span> <span class='kw'>=</span> <span class='fl'>5</span>,
  <span class='kw'>motif.search.window.size</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>motif.search.p.cutoff</span> <span class='kw'>=</span> <span class='fl'>1e-08</span>,
  <span class='kw'>cores</span> <span class='kw'>=</span> <span class='fl'>1</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>region</th>
      <td><p>A Granges or a named vector with
regions (i.e "chr21:100002-1004000")</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>Human genome reference "hg38" or "hg19"</p></td>
    </tr>
    <tr>
      <th>target.method</th>
      <td><p>How genes are mapped to regions: regions
overlapping gene promoter ("genes.promoter.overlap");
genes within a window around the region ("window"); or fixed number of
nearby genes upstream and
downstream from the region</p></td>
    </tr>
    <tr>
      <th>target.window.size</th>
      <td><p>When <code>method = "window"</code>, number of base
pairs to extend the region (+- window.size/2).
Default is 500kbp (or +/- 250kbp, i.e. 250k bp from start or end of the region)</p></td>
    </tr>
    <tr>
      <th>target.num.flanking.genes</th>
      <td><p>Number of flanking genes upstream and
downstream to search.
For example, if <code>target.num.flanking.genes = 5</code>, it will return the
5 genes upstream and 5 genes downstream</p></td>
    </tr>
    <tr>
      <th>motif.search.window.size</th>
      <td><p>Integer value to extend the regions.
For example, a value of 50 will
extend 25 bp upstream and 25 downstream the region. Default is no increase</p></td>
    </tr>
    <tr>
      <th>motif.search.p.cutoff</th>
      <td><p>motifmatchr pvalue cut-off. Default 1e-8.</p></td>
    </tr>
    <tr>
      <th>cores</th>
      <td><p>Number of CPU cores to be used. Default 1.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A data frame with TF, target and RegionID information.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='st'>"dna.met.chr21"</span>)
<span class='no'>dnam.regions</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='make_se_from_dnam_probes.html'>make_se_from_dnam_probes</a></span>(<span class='no'>dna.met.chr21</span>)
<span class='no'>triplet</span> <span class='kw'>&lt;-</span> <span class='fu'>create_triplet_distance_based</span>(
   <span class='kw'>region</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span>(<span class='no'>dnam.regions</span>)[<span class='fl'>1</span>:<span class='fl'>100</span>],
   <span class='kw'>motif.search.window.size</span> <span class='kw'>=</span> <span class='fl'>50</span>
)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Tiago Silva, Lily Wang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


