<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Fits linear models with interaction to triplet data (Target, TF, DNAm), where DNAm
can be a continuous variable (for all samples) or a binary variable (samples in Q1 or Q4) — interaction_model • MethReg</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Fits linear models with interaction to triplet data (Target, TF, DNAm), where DNAm
can be a continuous variable (for all samples) or a binary variable (samples in Q1 or Q4) — interaction_model" />
<meta property="og:description" content="Evaluates regulatory potential of DNA methylation (DNAm) on gene expression,
by fitting robust linear model or zero inflated negative binomial model to triplet data.
These models consist of terms to model direct effect of DNAm on target gene expression,
direct effect of TF on gene expression, as well as an interaction term that evaluates
the synergistic effect of DNAm and TF on gene expression." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MethReg</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/MethReg.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fits linear models with interaction to triplet data (Target, TF, DNAm), where DNAm
can be a continuous variable (for all samples) or a binary variable (samples in Q1 or Q4)</h1>
    
    <div class="hidden name"><code>interaction_model.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Evaluates regulatory potential of DNA methylation (DNAm) on gene expression,
by fitting robust linear model or zero inflated negative binomial model to triplet data.
These models consist of terms to model direct effect of DNAm on target gene expression,
direct effect of TF on gene expression, as well as an interaction term that evaluates
the synergistic effect of DNAm and TF on gene expression.</p>
    </div>

    <pre class="usage"><span class='fu'>interaction_model</span>(
  <span class='no'>triplet</span>,
  <span class='no'>dnam</span>,
  <span class='no'>exp</span>,
  <span class='kw'>cores</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>tf.activity.es</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>sig.threshold</span> <span class='kw'>=</span> <span class='fl'>0.05</span>,
  <span class='kw'>fdr</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>filter.correlated.tf.exp.dna</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>triplet</th>
      <td><p>Data frame with columns for DNA methylation region (regionID), TF  (TF), and target gene  (target)</p></td>
    </tr>
    <tr>
      <th>dnam</th>
      <td><p>DNA methylation matrix  (columns: samples in the same order as <code>exp</code> matrix, rows: regions/probes)</p></td>
    </tr>
    <tr>
      <th>exp</th>
      <td><p>A log2 (gene expression count + 1) matrix (columns: samples in the same order as <code>dnam</code> matrix,
rows: genes represented by ensembl IDs (e.g. ENSG00000239415))</p></td>
    </tr>
    <tr>
      <th>cores</th>
      <td><p>Number of CPU cores to be used. Default 1.</p></td>
    </tr>
    <tr>
      <th>tf.activity.es</th>
      <td><p>A matrix with normalized enrichment scores for each TF across all samples
to be used in linear models instead of TF gene expression.</p></td>
    </tr>
    <tr>
      <th>sig.threshold</th>
      <td><p>Threshold to filter significant triplets.
Select if interaction.pval &lt; 0.05 or pval.dnam &lt;0.05 or pval.tf &lt; 0.05 in binary model</p></td>
    </tr>
    <tr>
      <th>fdr</th>
      <td><p>Uses fdr when using sig.threshold.
Select if interaction.fdr &lt; 0.05 or fdr.dnam &lt;0.05 or fdr.tf &lt; 0.05 in binary model</p></td>
    </tr>
    <tr>
      <th>filter.correlated.tf.exp.dna</th>
      <td><p>If wilxocon test of TF expression Q1 and Q4 is significant (pvalue &lt; 0.05),
triplet will be removed.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A dataframe with <code>Region, TF, target, TF_symbo, target_symbol, estimates and P-values</code>,
after fitting robust linear models or zero-inflated negative binomial models (see Details above).</p>
<p>Model 1 (considering DNAm values as a continuous variable) generates <code>pval_met</code>, <code>pval_rna.tf</code>,
<code>pval_met.rna.tf</code> and <code>estimates_met</code>, <code>estimates_rna.tf</code>, <code>estimates_met.rna.tf</code>.</p>
<p>Model 2 (considering DNAm values as a binary variable) generates <code>quant_pval_metGrp</code>,
<code>quant_pval_rna.tf</code>, <code>quant_pval_metGrp.rna.tf</code>,
<code>quant_estimates_metGrp</code>, <code>quant_estimates_rna.tf</code>, <code>quant_estimates_metGrp.rna.tf</code>.</p>
<p><code>Model.interaction</code> indicates which model (robust linear model or zero inflated model)
was used to fit Model 1, and <code>Model.quantile</code> indicates which model(robust linear model or zero
inflated model) was used to fit Model 2.</p>
<p><code>Max_interaction_pval</code> is the max(quant_estimate_metGrp:rna.tf, pval_met:rna.tf).</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function fits the linear model</p>
<p><code>log2(RNA target) ~ log2(TF) + DNAm + log2(TF) * DNAm</code></p>
<p>to triplet data in two ways:</p>
<p>Model 1 : by considering <code>DNAm</code> as a continuous variable</p>
<p>Model 2:  by considering <code>DNAm</code> as a binary variable - we defined a binary group for
DNA methylation values (high = 1, low = 0). That is, samples with the highest
DNAm levels (top 25 percent) has high = 1, samples with lowest
DNAm levels (bottom 25 percent) has high = 0. Note that in this
implementation, only samples with DNAm values in the first and last quartiles
are considered.</p>
<p>In these models, the term <code><a href='https://rdrr.io/r/base/Log.html'>log2(TF)</a></code> evaluates direct effect of TF on
target gene expression, <code>DNAm</code> evaluates direct effect of DNAm on target
gene expression, and <code>log2(TF)*DNAm</code> evaluates synergistic effect of DNAm
and TF, that is, if TF regulatory activity is modified by DNAm.</p>
<p>There are two implementations of these models, depending on whether there are an excessive
amount (i.e. more than 25 percent) of samples with zero counts in RNAseq data:</p>
<ul>
<li><p>When percent of zeros in RNAseq data is less than
25 percent, robust linear models are implemented using <code>rlm</code> function from <code>MASS</code> package. This
gives outlier gene expression values reduced weight. We used <code>"psi.bisqure"</code>
option in function <code>rlm</code> (bisquare weighting,
https://stats.idre.ucla.edu/r/dae/robust-regression/).</p></li>
<li><p>When percent of zeros in RNAseq data is more than 25 percent, zero inflated negative binomial models
are implemented using <code>zeroinfl</code> function from <code>pscl</code> package. This assumes there are
two processes that generated zeros (1) one where the counts are always zero
(2) another where the count follows a negative binomial distribution.</p></li>
</ul>

<p>To account for confounding effects from covariate variables, first use the <code>get_residuals</code> function to obtain
RNA or DNAm residual values which have covariate effects removed, then fit interaction model. Note that no
log2 transformation is needed when <code>interaction_model</code> is applied to residuals data.</p>
<p>Note that only triplets with TF expression not significantly different in high vs. low
methylation groups will be evaluated (Wilcoxon test, p &gt; 0.05).</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/library.html'>library</a></span>(<span class='no'>dplyr</span>)
<span class='no'>dnam</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/Uniform.html'>runif</a></span>(<span class='fl'>20</span>,<span class='kw'>min</span> <span class='kw'>=</span> <span class='fl'>0</span>,<span class='kw'>max</span> <span class='kw'>=</span> <span class='fl'>1</span>) <span class='kw'>%&gt;%</span>
  <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>matrix</a></span>(<span class='kw'>ncol</span> <span class='kw'>=</span> <span class='fl'>1</span>) <span class='kw'>%&gt;%</span>  <span class='no'>t</span>
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span>(<span class='no'>dnam</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"chr3:203727581-203728580"</span>)
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span>(<span class='no'>dnam</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span>(<span class='st'>"Samples"</span>,<span class='fl'>1</span>:<span class='fl'>20</span>)

<span class='no'>exp.target</span> <span class='kw'>&lt;-</span>  <span class='fu'><a href='https://rdrr.io/r/stats/Uniform.html'>runif</a></span>(<span class='fl'>20</span>,<span class='kw'>min</span> <span class='kw'>=</span> <span class='fl'>0</span>,<span class='kw'>max</span> <span class='kw'>=</span> <span class='fl'>10</span>) <span class='kw'>%&gt;%</span>
  <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>matrix</a></span>(<span class='kw'>ncol</span> <span class='kw'>=</span> <span class='fl'>1</span>) <span class='kw'>%&gt;%</span>  <span class='no'>t</span>
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span>(<span class='no'>exp.target</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"ENSG00000232886"</span>)
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span>(<span class='no'>exp.target</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span>(<span class='st'>"Samples"</span>,<span class='fl'>1</span>:<span class='fl'>20</span>)

<span class='no'>exp.tf</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/Uniform.html'>runif</a></span>(<span class='fl'>20</span>,<span class='kw'>min</span> <span class='kw'>=</span> <span class='fl'>0</span>,<span class='kw'>max</span> <span class='kw'>=</span> <span class='fl'>10</span>) <span class='kw'>%&gt;%</span>
  <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>matrix</a></span>(<span class='kw'>ncol</span> <span class='kw'>=</span> <span class='fl'>1</span>) <span class='kw'>%&gt;%</span>  <span class='no'>t</span>
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span>(<span class='no'>exp.tf</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"ENSG00000232888"</span>)
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span>(<span class='no'>exp.tf</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span>(<span class='st'>"Samples"</span>,<span class='fl'>1</span>:<span class='fl'>20</span>)

<span class='no'>exp</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>rbind</a></span>(<span class='no'>exp.tf</span>, <span class='no'>exp.target</span>)

<span class='no'>triplet</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span>(
   <span class='st'>"regionID"</span> <span class='kw'>=</span>  <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"chr3:203727581-203728580"</span>),
   <span class='st'>"target"</span> <span class='kw'>=</span> <span class='st'>"ENSG00000232886"</span>,
   <span class='st'>"TF"</span> <span class='kw'>=</span> <span class='st'>"ENSG00000232888"</span>
)
<span class='no'>results</span> <span class='kw'>&lt;-</span> <span class='fu'>interaction_model</span>(<span class='no'>triplet</span>, <span class='no'>dnam</span>, <span class='no'>exp</span>)</div><div class='output co'>#&gt; <span class='message'>Removing genes with RNA expression equal to 0 for all samples from triplets</span></div><div class='output co'>#&gt; <span class='message'>Removing triplet with no DNA methylation information for more than 25% of the samples</span></div><div class='output co'>#&gt; |                                                    |  0%                      |====================================================|100% ~0 s remaining       |====================================================|100%                      Completed after 0 s </div><div class='output co'>#&gt; <span class='message'>Filtering results to have interaction, TF or DNAm significant</span></div><div class='output co'>#&gt; <span class='message'>Filtering results to wilcoxon test TF Q1 vs Q4 not significant</span></div><div class='input'>
<span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='st'>"dna.met.chr21"</span>)
<span class='no'>dna.met.chr21</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='make_se_from_dnam_probes.html'>make_se_from_dnam_probes</a></span>(<span class='no'>dna.met.chr21</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='st'>"gene.exp.chr21.log2"</span>)
<span class='no'>triplet</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span>(
    <span class='st'>"regionID"</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span>(<span class='no'>dna.met.chr21</span>)[<span class='fl'>1</span>:<span class='fl'>10</span>],
    <span class='st'>"TF"</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span>(<span class='no'>gene.exp.chr21.log2</span>)[<span class='fl'>11</span>:<span class='fl'>20</span>],
    <span class='st'>"target"</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span>(<span class='no'>gene.exp.chr21.log2</span>)[<span class='fl'>1</span>:<span class='fl'>10</span>]
)
<span class='no'>results</span> <span class='kw'>&lt;-</span> <span class='fu'>interaction_model</span>(<span class='no'>triplet</span>, <span class='no'>dna.met.chr21</span>, <span class='no'>gene.exp.chr21.log2</span>)
<span class='co'># select those that are significant in both models</span>
<span class='co'># results &lt;- results[results$`pval_met:rna.tf`&lt; 0.05 &amp; results$`quant_pval_metGrp:rna.tf`&lt; 0.05 ,]</span>
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Tiago Silva, Lily Wang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


