<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Fits linear models to triplet data (Target, TF, DNAm) for
samples with high DNAm or low DNAm separately, and annotates TF (activator/repressor,
methyl-plus/methyl-minus). — stratified_model • MethReg</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Fits linear models to triplet data (Target, TF, DNAm) for
samples with high DNAm or low DNAm separately, and annotates TF (activator/repressor,
methyl-plus/methyl-minus). — stratified_model" />
<meta property="og:description" content="Should be used after fitting interaction_model, and only
for triplet data with significant TF*DNAm interaction. This analysis
examines in more details on how TF activities differ in samples with high DNAm or low DNAm values." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MethReg</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/MethReg.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fits linear models to triplet data (Target, TF, DNAm) for
samples with high DNAm or low DNAm separately, and annotates TF (activator/repressor,
methyl-plus/methyl-minus).</h1>
    
    <div class="hidden name"><code>stratified_model.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Should be used after fitting <code>interaction_model</code>, and only
for triplet data with significant <code>TF*DNAm</code> interaction. This analysis
examines in more details on how TF activities differ in samples with high DNAm or low DNAm values.</p>
    </div>

    <pre class="usage"><span class='fu'>stratified_model</span>(<span class='no'>triplet</span>, <span class='no'>dnam</span>, <span class='no'>exp</span>, <span class='kw'>cores</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>tf.activity.es</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>triplet</th>
      <td><p>Data frame with columns for DNA methylation region (regionID), TF  (TF), and target gene  (target)</p></td>
    </tr>
    <tr>
      <th>dnam</th>
      <td><p>DNA methylation matrix  (columns: samples in the same order as <code>exp</code> matrix, rows: regions/probes)</p></td>
    </tr>
    <tr>
      <th>exp</th>
      <td><p>A log2 (gene expression + 1) matrix (columns: samples in the same order as <code>dnam</code> matrix,
rows: genes represented by ensembl IDs (e.g. ENSG00000239415))</p></td>
    </tr>
    <tr>
      <th>cores</th>
      <td><p>Number of CPU cores to be used. Default 1.</p></td>
    </tr>
    <tr>
      <th>tf.activity.es</th>
      <td><p>A matrix with normalized enrichment scores for each TF across all samples
to be used in linear models instead of TF gene expression.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A dataframe with <code>Region, TF, target, TF_symbol target_symbol</code>, results for
fitting linear models to samples with low methylation (<code>DNAmlow_pval_rna.tf</code>,
<code>DNAmlow_estimate_rna.tf</code>), or samples with high methylation (<code>DNAmhigh_pval_rna.tf</code>,
<code>DNAmhigh_pval_rna.tf.1</code>), annotations for TF (<code>class.TF</code>) and (<code>class.TF.DNAm</code>).</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function fits linear model
<code>log2(RNA target) = log2(TF)</code></p>
<p>to samples with highest DNAm values (top 25 percent) or lowest DNAm values (bottom 25 percent), separately.</p>
<p>There are two implementations of these models, depending on whether there are an excessive
amount (i.e. more than 25 percent) of samples with zero counts in RNAseq data:</p>
<ul>
<li><p>When percent of zeros in RNAseq data is less than
25 percent, robust linear models are implemented using <code>rlm</code> function from <code>MASS</code> package. This
gives outlier gene expression values reduced weight. We used <code>"psi.bisqure"</code>
option in function <code>rlm</code> (bisquare weighting,
https://stats.idre.ucla.edu/r/dae/robust-regression/).</p></li>
<li><p>When percent of zeros in RNAseq data is more than 25 percent, zero inflated negative binomial models
are implemented using <code>zeroinfl</code> function from <code>pscl</code> package. This assumes there are
two processes that generated zeros (1) one where the counts are always zero
(2) another where the count follows a negative binomial distribution.</p></li>
</ul>

<p>To account for confounding effects from covariate variables, first use the <code>get_residuals</code> function to obtain
RNA residual values which have covariate effects removed, then fit interaction model. Note that no
log2 transformation is needed when <code>interaction_model</code> is applied to residuals data.</p>
<p>This function also provides annotations for TFs. A TF is annotated as <code>activator</code> if
increasing amount of TF (higher TF gene expression) corresponds to increased target gene expression. A TF
is annotated as <code>repressor</code> if increasing amount of TF (higher TF gene expression) corresponds to
decrease in target gene expression.</p>
<p>In addition, a region/CpG is annotated as <code>enhancing</code> if more TF regulation on gene transcription
is observed in samples with high DNAm. That is,  DNA methylation enhances TF regulation on target gene expression.
On the other hand, a region/CpG is annotated as <code>attenuating</code> (methyl-minus) if more TF regulation on gene
transcription is observed in samples with low DNAm. That is, DNA methylation reduces TF regulation
on target gene expression.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/library.html'>library</a></span>(<span class='no'>dplyr</span>)
<span class='no'>dnam</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/Uniform.html'>runif</a></span>(<span class='fl'>20</span>,<span class='kw'>min</span> <span class='kw'>=</span> <span class='fl'>0</span>,<span class='kw'>max</span> <span class='kw'>=</span> <span class='fl'>1</span>) <span class='kw'>%&gt;%</span>
  <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>matrix</a></span>(<span class='kw'>ncol</span> <span class='kw'>=</span> <span class='fl'>1</span>) <span class='kw'>%&gt;%</span>  <span class='no'>t</span>
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span>(<span class='no'>dnam</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"chr3:203727581-203728580"</span>)
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span>(<span class='no'>dnam</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span>(<span class='st'>"Samples"</span>,<span class='fl'>1</span>:<span class='fl'>20</span>)

<span class='no'>exp.target</span> <span class='kw'>&lt;-</span>  <span class='fu'><a href='https://rdrr.io/r/stats/Uniform.html'>runif</a></span>(<span class='fl'>20</span>,<span class='kw'>min</span> <span class='kw'>=</span> <span class='fl'>0</span>,<span class='kw'>max</span> <span class='kw'>=</span> <span class='fl'>10</span>) <span class='kw'>%&gt;%</span>
  <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>matrix</a></span>(<span class='kw'>ncol</span> <span class='kw'>=</span> <span class='fl'>1</span>) <span class='kw'>%&gt;%</span>  <span class='no'>t</span>
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span>(<span class='no'>exp.target</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"ENSG00000232886"</span>)
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span>(<span class='no'>exp.target</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span>(<span class='st'>"Samples"</span>,<span class='fl'>1</span>:<span class='fl'>20</span>)

<span class='no'>exp.tf</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/Uniform.html'>runif</a></span>(<span class='fl'>20</span>,<span class='kw'>min</span> <span class='kw'>=</span> <span class='fl'>0</span>,<span class='kw'>max</span> <span class='kw'>=</span> <span class='fl'>10</span>) <span class='kw'>%&gt;%</span>
  <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>matrix</a></span>(<span class='kw'>ncol</span> <span class='kw'>=</span> <span class='fl'>1</span>) <span class='kw'>%&gt;%</span>  <span class='no'>t</span>
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span>(<span class='no'>exp.tf</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"ENSG00000232888"</span>)
<span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span>(<span class='no'>exp.tf</span>) <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span>(<span class='st'>"Samples"</span>,<span class='fl'>1</span>:<span class='fl'>20</span>)

<span class='no'>exp</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>rbind</a></span>(<span class='no'>exp.tf</span>, <span class='no'>exp.target</span>)

<span class='no'>triplet</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span>(
   <span class='st'>"regionID"</span> <span class='kw'>=</span>  <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"chr3:203727581-203728580"</span>),
   <span class='st'>"target"</span> <span class='kw'>=</span> <span class='st'>"ENSG00000232886"</span>,
   <span class='st'>"TF"</span> <span class='kw'>=</span> <span class='st'>"ENSG00000232888"</span>
)

<span class='no'>results</span> <span class='kw'>&lt;-</span> <span class='fu'>stratified_model</span>(<span class='kw'>triplet</span> <span class='kw'>=</span> <span class='no'>triplet</span>,<span class='kw'>dnam</span> <span class='kw'>=</span> <span class='no'>dnam</span>, <span class='kw'>exp</span> <span class='kw'>=</span> <span class='no'>exp</span>)</div><div class='output co'>#&gt; <span class='message'>Removing genes with RNA expression equal to 0 for all samples from triplets</span></div><div class='output co'>#&gt; <span class='message'>Removing triplet with no DNA methylation information for more than 25% of the samples</span></div><div class='output co'>#&gt; |                                                    |  0%                      |====================================================|100% ~0 s remaining       |====================================================|100%                      Completed after 0 s </div><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='st'>"dna.met.chr21"</span>)
<span class='no'>dna.met.chr21</span> <span class='kw'>&lt;-</span> <span class='fu'>map_probes_to_regions</span>(<span class='no'>dna.met.chr21</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='st'>"gene.exp.chr21.log2"</span>)
<span class='no'>triplet</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span>(<span class='st'>"regionID"</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span>(<span class='no'>dna.met.chr21</span>)[<span class='fl'>1</span>:<span class='fl'>10</span>],
                      <span class='st'>"TF"</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span>(<span class='no'>gene.exp.chr21.log2</span>)[<span class='fl'>11</span>:<span class='fl'>20</span>],
                      <span class='st'>"target"</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span>(<span class='no'>gene.exp.chr21.log2</span>)[<span class='fl'>1</span>:<span class='fl'>10</span>])
<span class='no'>results</span> <span class='kw'>&lt;-</span> <span class='fu'>stratified_model</span>(<span class='no'>triplet</span>, <span class='no'>dna.met.chr21</span>, <span class='no'>gene.exp.chr21.log2</span>)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Tiago Silva, Lily Wang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


